// @hyperdx/otel-web-session-recorder/dist/esm/opentelemetry/proto/resource/v1/resource.js
/* eslint-disable */
import _m0 from 'protobufjs/minimal';
import { KeyValue } from '../../common/v1/common';
export const protobufPackage = 'opentelemetry.proto.resource.v1';
function createBaseResource() {
    return { attributes: [], droppedAttributesCount: 0 };
}
export const Resource = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.attributes !== undefined && message.attributes.length !== 0) {
            for (const v of message.attributes) {
                KeyValue.encode(v, writer.uint32(10).fork()).ldelim();
            }
        }
        if (message.droppedAttributesCount !== undefined &&
            message.droppedAttributesCount !== 0) {
            writer.uint32(16).uint32(message.droppedAttributesCount);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseResource();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.attributes.push(KeyValue.decode(reader, reader.uint32()));
                    continue;
                case 2:
                    if (tag !== 16) {
                        break;
                    }
                    message.droppedAttributesCount = reader.uint32();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            attributes: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.attributes)
                ? object.attributes.map((e) => KeyValue.fromJSON(e))
                : [],
            droppedAttributesCount: isSet(object.droppedAttributesCount)
                ? globalThis.Number(object.droppedAttributesCount)
                : 0,
        };
    },
    toJSON(message) {
        var _a;
        const obj = {};
        if ((_a = message.attributes) === null || _a === void 0 ? void 0 : _a.length) {
            obj.attributes = message.attributes.map((e) => KeyValue.toJSON(e));
        }
        if (message.droppedAttributesCount !== undefined &&
            message.droppedAttributesCount !== 0) {
            obj.droppedAttributesCount = Math.round(message.droppedAttributesCount);
        }
        return obj;
    },
    create(base) {
        return Resource.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b;
        const message = createBaseResource();
        message.attributes =
            ((_a = object.attributes) === null || _a === void 0 ? void 0 : _a.map((e) => KeyValue.fromPartial(e))) || [];
        message.droppedAttributesCount = (_b = object.droppedAttributesCount) !== null && _b !== void 0 ? _b : 0;
        return message;
    },
};
function isSet(value) {
    return value !== null && value !== undefined;
}
