// @hyperdx/instrumentation-exception/build/src/node/integrations/context.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getDeviceContext = exports.getAppContext = exports.nodeContextIntegration = exports._nodeContextIntegration = exports.readDirAsync = exports.readFileAsync = void 0;
const tslib_1 = require("tslib");
const node_child_process_1 = require("node:child_process");
const node_fs_1 = require("node:fs");
const os = tslib_1.__importStar(require("node:os"));
const node_path_1 = require("node:path");
const node_util_1 = require("node:util");
const core_1 = require("@sentry/core");
exports.readFileAsync = (0, node_util_1.promisify)(node_fs_1.readFile);
exports.readDirAsync = (0, node_util_1.promisify)(node_fs_1.readdir);
const INTEGRATION_NAME = 'Context';
exports._nodeContextIntegration = ((options = {}) => {
    let cachedContext;
    const _options = {
        app: true,
        os: true,
        device: true,
        culture: true,
        cloudResource: true,
        ...options,
    };
    async function addContext(event) {
        var _a, _b, _c, _d, _e;
        if (cachedContext === undefined) {
            cachedContext = _getContexts();
        }
        const updatedContext = _updateContext(await cachedContext);
        event.contexts = {
            ...event.contexts,
            app: { ...updatedContext.app, ...(_a = event.contexts) === null || _a === void 0 ? void 0 : _a.app },
            os: { ...updatedContext.os, ...(_b = event.contexts) === null || _b === void 0 ? void 0 : _b.os },
            device: { ...updatedContext.device, ...(_c = event.contexts) === null || _c === void 0 ? void 0 : _c.device },
            culture: { ...updatedContext.culture, ...(_d = event.contexts) === null || _d === void 0 ? void 0 : _d.culture },
            cloud_resource: {
                ...updatedContext.cloud_resource,
                ...(_e = event.contexts) === null || _e === void 0 ? void 0 : _e.cloud_resource,
            },
        };
        return event;
    }
    async function _getContexts() {
        const contexts = {};
        if (_options.os) {
            contexts.os = await getOsContext();
        }
        if (_options.app) {
            contexts.app = getAppContext();
        }
        if (_options.device) {
            contexts.device = getDeviceContext(_options.device);
        }
        if (_options.culture) {
            const culture = getCultureContext();
            if (culture) {
                contexts.culture = culture;
            }
        }
        if (_options.cloudResource) {
            contexts.cloud_resource = getCloudResourceContext();
        }
        return contexts;
    }
    return {
        name: INTEGRATION_NAME,
        processEvent(event) {
            return addContext(event);
        },
    };
});
exports.nodeContextIntegration = (0, core_1.defineIntegration)(exports._nodeContextIntegration);
function _updateContext(contexts) {
    var _a, _b, _c, _d, _e;
    if ((_a = contexts === null || contexts === void 0 ? void 0 : contexts.app) === null || _a === void 0 ? void 0 : _a.app_memory) {
        contexts.app.app_memory = process.memoryUsage().rss;
    }
    if (((_b = contexts === null || contexts === void 0 ? void 0 : contexts.app) === null || _b === void 0 ? void 0 : _b.free_memory) &&
        typeof process.availableMemory === 'function') {
        const freeMemory = (_d = (_c = process).availableMemory) === null || _d === void 0 ? void 0 : _d.call(_c);
        if (freeMemory != null) {
            contexts.app.free_memory = freeMemory;
        }
    }
    if ((_e = contexts === null || contexts === void 0 ? void 0 : contexts.device) === null || _e === void 0 ? void 0 : _e.free_memory) {
        contexts.device.free_memory = os.freemem();
    }
    return contexts;
}
async function getOsContext() {
    const platformId = os.platform();
    switch (platformId) {
        case 'darwin':
            return getDarwinInfo();
        case 'linux':
            return getLinuxInfo();
        default:
            return {
                name: PLATFORM_NAMES[platformId] || platformId,
                version: os.release(),
            };
    }
}
function getCultureContext() {
    try {
        if (typeof process.versions.icu !== 'string') {
            return;
        }
        const january = new Date(9e8);
        const spanish = new Intl.DateTimeFormat('es', { month: 'long' });
        if (spanish.format(january) === 'enero') {
            const options = Intl.DateTimeFormat().resolvedOptions();
            return {
                locale: options.locale,
                timezone: options.timeZone,
            };
        }
    }
    catch (err) {
    }
    return;
}
function getAppContext() {
    var _a, _b;
    const app_memory = process.memoryUsage().rss;
    const app_start_time = new Date(Date.now() - process.uptime() * 1000).toISOString();
    const appContext = { app_start_time, app_memory };
    if (typeof process.availableMemory === 'function') {
        const freeMemory = (_b = (_a = process).availableMemory) === null || _b === void 0 ? void 0 : _b.call(_a);
        if (freeMemory != null) {
            appContext.free_memory = freeMemory;
        }
    }
    return appContext;
}
exports.getAppContext = getAppContext;
function getDeviceContext(deviceOpt) {
    const device = {};
    let uptime;
    try {
        uptime = os.uptime && os.uptime();
    }
    catch (e) {
    }
    if (typeof uptime === 'number') {
        device.boot_time = new Date(Date.now() - uptime * 1000).toISOString();
    }
    device.arch = os.arch();
    if (deviceOpt === true || deviceOpt.memory) {
        device.memory_size = os.totalmem();
        device.free_memory = os.freemem();
    }
    if (deviceOpt === true || deviceOpt.cpu) {
        const cpuInfo = os.cpus();
        if (cpuInfo && cpuInfo.length) {
            const firstCpu = cpuInfo[0];
            device.processor_count = cpuInfo.length;
            device.cpu_description = firstCpu.model;
            device.processor_frequency = firstCpu.speed;
        }
    }
    return device;
}
exports.getDeviceContext = getDeviceContext;
const PLATFORM_NAMES = {
    aix: 'IBM AIX',
    freebsd: 'FreeBSD',
    openbsd: 'OpenBSD',
    sunos: 'SunOS',
    win32: 'Windows',
};
const LINUX_DISTROS = [
    { name: 'fedora-release', distros: ['Fedora'] },
    { name: 'redhat-release', distros: ['Red Hat Linux', 'Centos'] },
    { name: 'redhat_version', distros: ['Red Hat Linux'] },
    { name: 'SuSE-release', distros: ['SUSE Linux'] },
    { name: 'lsb-release', distros: ['Ubuntu Linux', 'Arch Linux'] },
    { name: 'debian_version', distros: ['Debian'] },
    { name: 'debian_release', distros: ['Debian'] },
    { name: 'arch-release', distros: ['Arch Linux'] },
    { name: 'gentoo-release', distros: ['Gentoo Linux'] },
    { name: 'novell-release', distros: ['SUSE Linux'] },
    { name: 'alpine-release', distros: ['Alpine Linux'] },
];
const LINUX_VERSIONS = {
    alpine: (content) => content,
    arch: (content) => matchFirst(/distrib_release=(.*)/, content),
    centos: (content) => matchFirst(/release ([^ ]+)/, content),
    debian: (content) => content,
    fedora: (content) => matchFirst(/release (..)/, content),
    mint: (content) => matchFirst(/distrib_release=(.*)/, content),
    red: (content) => matchFirst(/release ([^ ]+)/, content),
    suse: (content) => matchFirst(/VERSION = (.*)\n/, content),
    ubuntu: (content) => matchFirst(/distrib_release=(.*)/, content),
};
function matchFirst(regex, text) {
    const match = regex.exec(text);
    return match ? match[1] : undefined;
}
async function getDarwinInfo() {
    const darwinInfo = {
        kernel_version: os.release(),
        name: 'Mac OS X',
        version: `10.${Number(os.release().split('.')[0]) - 4}`,
    };
    try {
        const output = await new Promise((resolve, reject) => {
            (0, node_child_process_1.execFile)('/usr/bin/sw_vers', (error, stdout) => {
                if (error) {
                    reject(error);
                    return;
                }
                resolve(stdout);
            });
        });
        darwinInfo.name = matchFirst(/^ProductName:\s+(.*)$/m, output);
        darwinInfo.version = matchFirst(/^ProductVersion:\s+(.*)$/m, output);
        darwinInfo.build = matchFirst(/^BuildVersion:\s+(.*)$/m, output);
    }
    catch (e) {
    }
    return darwinInfo;
}
function getLinuxDistroId(name) {
    return name.split(' ')[0].toLowerCase();
}
async function getLinuxInfo() {
    const linuxInfo = {
        kernel_version: os.release(),
        name: 'Linux',
    };
    try {
        const etcFiles = await (0, exports.readDirAsync)('/etc');
        const distroFile = LINUX_DISTROS.find((file) => etcFiles.includes(file.name));
        if (!distroFile) {
            return linuxInfo;
        }
        const distroPath = (0, node_path_1.join)('/etc', distroFile.name);
        const contents = (await (0, exports.readFileAsync)(distroPath, { encoding: 'utf-8' })).toLowerCase();
        const { distros } = distroFile;
        linuxInfo.name =
            distros.find((d) => contents.indexOf(getLinuxDistroId(d)) >= 0) ||
                distros[0];
        const id = getLinuxDistroId(linuxInfo.name);
        linuxInfo.version = LINUX_VERSIONS[id](contents);
    }
    catch (e) {
    }
    return linuxInfo;
}
function getCloudResourceContext() {
    if (process.env.VERCEL) {
        return {
            'cloud.provider': 'vercel',
            'cloud.region': process.env.VERCEL_REGION,
        };
    }
    else if (process.env.AWS_REGION) {
        return {
            'cloud.provider': 'aws',
            'cloud.region': process.env.AWS_REGION,
            'cloud.platform': process.env.AWS_EXECUTION_ENV,
        };
    }
    else if (process.env.GCP_PROJECT) {
        return {
            'cloud.provider': 'gcp',
        };
    }
    else if (process.env.ALIYUN_REGION_ID) {
        return {
            'cloud.provider': 'alibaba_cloud',
            'cloud.region': process.env.ALIYUN_REGION_ID,
        };
    }
    else if (process.env.WEBSITE_SITE_NAME && process.env.REGION_NAME) {
        return {
            'cloud.provider': 'azure',
            'cloud.region': process.env.REGION_NAME,
        };
    }
    else if (process.env.IBM_CLOUD_REGION) {
        return {
            'cloud.provider': 'ibm_cloud',
            'cloud.region': process.env.IBM_CLOUD_REGION,
        };
    }
    else if (process.env.TENCENTCLOUD_REGION) {
        return {
            'cloud.provider': 'tencent_cloud',
            'cloud.region': process.env.TENCENTCLOUD_REGION,
            'cloud.account.id': process.env.TENCENTCLOUD_APPID,
            'cloud.availability_zone': process.env.TENCENTCLOUD_ZONE,
        };
    }
    else if (process.env.NETLIFY) {
        return {
            'cloud.provider': 'netlify',
        };
    }
    else if (process.env.FLY_REGION) {
        return {
            'cloud.provider': 'fly.io',
            'cloud.region': process.env.FLY_REGION,
        };
    }
    else if (process.env.DYNO) {
        return {
            'cloud.provider': 'heroku',
        };
    }
    else {
        return undefined;
    }
}
