// @hyperdx/instrumentation-exception/build/src/node/integrations/contextlines.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.contextLinesIntegration = exports._contextLinesIntegration = exports.resetFileContentCache = void 0;
const node_fs_1 = require("node:fs");
const core_1 = require("@sentry/core");
const utils_1 = require("@sentry/utils");
const FILE_CONTENT_CACHE = new utils_1.LRUMap(100);
const DEFAULT_LINES_OF_CONTEXT = 7;
const INTEGRATION_NAME = 'ContextLines';
const readFileAsync = node_fs_1.promises.readFile;
function resetFileContentCache() {
    FILE_CONTENT_CACHE.clear();
}
exports.resetFileContentCache = resetFileContentCache;
exports._contextLinesIntegration = ((options = {}) => {
    const contextLines = options.frameContextLines !== undefined
        ? options.frameContextLines
        : DEFAULT_LINES_OF_CONTEXT;
    return {
        name: INTEGRATION_NAME,
        processEvent(event) {
            return addSourceContext(event, contextLines);
        },
    };
});
exports.contextLinesIntegration = (0, core_1.defineIntegration)(exports._contextLinesIntegration);
async function addSourceContext(event, contextLines) {
    var _a, _b, _c;
    const enqueuedReadSourceFileTasks = {};
    const readSourceFileTasks = [];
    if (contextLines > 0 && ((_a = event.exception) === null || _a === void 0 ? void 0 : _a.values)) {
        for (const exception of event.exception.values) {
            if (!((_b = exception.stacktrace) === null || _b === void 0 ? void 0 : _b.frames)) {
                continue;
            }
            for (let i = exception.stacktrace.frames.length - 1; i >= 0; i--) {
                const frame = exception.stacktrace.frames[i];
                if (frame.filename &&
                    !enqueuedReadSourceFileTasks[frame.filename] &&
                    !FILE_CONTENT_CACHE.get(frame.filename)) {
                    readSourceFileTasks.push(_readSourceFile(frame.filename));
                    enqueuedReadSourceFileTasks[frame.filename] = 1;
                }
            }
        }
    }
    if (readSourceFileTasks.length > 0) {
        await Promise.all(readSourceFileTasks);
    }
    if (contextLines > 0 && ((_c = event.exception) === null || _c === void 0 ? void 0 : _c.values)) {
        for (const exception of event.exception.values) {
            if (exception.stacktrace && exception.stacktrace.frames) {
                await addSourceContextToFrames(exception.stacktrace.frames, contextLines);
            }
        }
    }
    return event;
}
function addSourceContextToFrames(frames, contextLines) {
    for (const frame of frames) {
        if (frame.filename && frame.context_line === undefined) {
            const sourceFileLines = FILE_CONTENT_CACHE.get(frame.filename);
            if (sourceFileLines) {
                try {
                    (0, utils_1.addContextToFrame)(sourceFileLines, frame, contextLines);
                }
                catch (e) {
                }
            }
        }
    }
}
async function _readSourceFile(filename) {
    const cachedFile = FILE_CONTENT_CACHE.get(filename);
    if (cachedFile === null) {
        return null;
    }
    if (cachedFile !== undefined) {
        return cachedFile;
    }
    let content = null;
    try {
        const rawFileContents = await readFileAsync(filename, 'utf-8');
        content = rawFileContents.split('\n');
    }
    catch (_) {
    }
    FILE_CONTENT_CACHE.set(filename, content);
    return content;
}
