// @hyperdx/instrumentation-exception/build/src/node/integrations/onunhandledrejection.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.makeUnhandledPromiseHandler = exports.onUnhandledRejectionIntegration = void 0;
const utils_1 = require("@sentry/utils");
const core_1 = require("@sentry/core");
const api_1 = require("@opentelemetry/api");
const errorhandling_1 = require("../utils/errorhandling");
const __1 = require("..");
const INTEGRATION_NAME = 'OnUnhandledRejection';
const _onUnhandledRejectionIntegration = ((options) => {
    const mode = options.mode || 'warn';
    return {
        name: INTEGRATION_NAME,
        setup() {
            global.process.on('unhandledRejection', makeUnhandledPromiseHandler({
                mode,
                forceFlush: options.forceFlush,
            }));
            api_1.diag.debug('Registered global error handler for unhandled promise rejections');
        },
    };
});
exports.onUnhandledRejectionIntegration = (0, core_1.defineIntegration)(_onUnhandledRejectionIntegration);
function makeUnhandledPromiseHandler(options) {
    return function sendUnhandledPromise(reason, promise) {
        const p = (0, __1.recordException)(reason, {
            originalException: promise,
            captureContext: {
                extra: { unhandledPromiseRejection: true },
            },
            mechanism: {
                handled: false,
                type: 'onunhandledrejection',
            },
        });
        handleRejection(reason, options, p);
    };
}
exports.makeUnhandledPromiseHandler = makeUnhandledPromiseHandler;
function handleRejection(reason, options, pendingPromise) {
    const rejectionWarning = 'This error originated either by ' +
        'throwing inside of an async function without a catch block, ' +
        'or by rejecting a promise which was not handled with .catch().' +
        ' The promise rejected with the reason:';
    if (options.mode === 'warn') {
        (0, utils_1.consoleSandbox)(() => {
            console.warn(rejectionWarning);
            console.error(reason && reason.stack ? reason.stack : reason);
        });
    }
    else if (options.mode === 'strict') {
        (0, utils_1.consoleSandbox)(() => {
            console.warn(rejectionWarning);
        });
        (pendingPromise !== null && pendingPromise !== void 0 ? pendingPromise : Promise.resolve()).finally(() => {
            console.log('Exiting due to unhandled promise rejection');
            (0, errorhandling_1.logAndExitProcess)(reason, options.forceFlush);
        });
    }
}
