// @hyperdx/instrumentation-exception/build/src/node/integrations/onuncaughtexception.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.makeErrorHandler = exports.onUncaughtExceptionIntegration = void 0;
const core_1 = require("@sentry/core");
const api_1 = require("@opentelemetry/api");
const errorhandling_1 = require("../utils/errorhandling");
const __1 = require("..");
const INTEGRATION_NAME = 'OnUncaughtException';
exports.onUncaughtExceptionIntegration = (0, core_1.defineIntegration)((options) => {
    const optionsWithDefaults = {
        exitEvenIfOtherHandlersAreRegistered: false,
        ...options,
    };
    return {
        name: INTEGRATION_NAME,
        setup() {
            global.process.on('uncaughtException', makeErrorHandler(optionsWithDefaults));
            api_1.diag.debug('Registered global error handler for uncaught exceptions');
        },
    };
});
function makeErrorHandler(options) {
    const timeout = 2000;
    let caughtFirstError = false;
    let caughtSecondError = false;
    let calledFatalError = false;
    let firstError;
    return Object.assign((error) => {
        const onFatalError = errorhandling_1.logAndExitProcess;
        const userProvidedListenersCount = global.process.listeners('uncaughtException').filter((listener) => {
            return (listener.name !== 'domainUncaughtExceptionClear' &&
                listener._hdxErrorHandler !== true);
        }).length;
        const processWouldExit = userProvidedListenersCount === 0;
        const shouldApplyFatalHandlingLogic = options.exitEvenIfOtherHandlersAreRegistered || processWouldExit;
        if (!caughtFirstError) {
            firstError = error;
            caughtFirstError = true;
            const p = (0, __1.recordException)(error, {
                originalException: error,
                captureContext: {
                    level: 'fatal',
                },
                mechanism: {
                    handled: false,
                    type: 'onuncaughtexception',
                },
            });
            if (!calledFatalError && shouldApplyFatalHandlingLogic) {
                calledFatalError = true;
                p.finally(() => {
                    onFatalError(error, options.forceFlush);
                });
            }
        }
        else {
            if (shouldApplyFatalHandlingLogic) {
                if (calledFatalError) {
                    api_1.diag.warn('uncaught exception after calling fatal error shutdown callback - this is bad! forcing shutdown');
                    (0, errorhandling_1.logAndExitProcess)(error, options.forceFlush);
                }
                else if (!caughtSecondError) {
                    caughtSecondError = true;
                    setTimeout(() => {
                        if (!calledFatalError) {
                            calledFatalError = true;
                            onFatalError(firstError, options.forceFlush);
                        }
                        else {
                        }
                    }, timeout);
                }
            }
        }
    }, { _hdxErrorHandler: true });
}
exports.makeErrorHandler = makeErrorHandler;
