// @hyperdx/otel-web-session-recorder/dist/esm/opentelemetry/proto/common/v1/common.js
/* eslint-disable */
import Long from 'long';
import _m0 from 'protobufjs/minimal';
export const protobufPackage = 'opentelemetry.proto.common.v1';
function createBaseAnyValue() {
    return {
        stringValue: undefined,
        boolValue: undefined,
        intValue: undefined,
        doubleValue: undefined,
        arrayValue: undefined,
        kvlistValue: undefined,
        bytesValue: undefined,
    };
}
export const AnyValue = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.stringValue !== undefined) {
            writer.uint32(10).string(message.stringValue);
        }
        if (message.boolValue !== undefined) {
            writer.uint32(16).bool(message.boolValue);
        }
        if (message.intValue !== undefined) {
            writer.uint32(24).int64(message.intValue);
        }
        if (message.doubleValue !== undefined) {
            writer.uint32(33).double(message.doubleValue);
        }
        if (message.arrayValue !== undefined) {
            ArrayValue.encode(message.arrayValue, writer.uint32(42).fork()).ldelim();
        }
        if (message.kvlistValue !== undefined) {
            KeyValueList.encode(message.kvlistValue, writer.uint32(50).fork()).ldelim();
        }
        if (message.bytesValue !== undefined) {
            writer.uint32(58).bytes(message.bytesValue);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseAnyValue();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.stringValue = reader.string();
                    continue;
                case 2:
                    if (tag !== 16) {
                        break;
                    }
                    message.boolValue = reader.bool();
                    continue;
                case 3:
                    if (tag !== 24) {
                        break;
                    }
                    message.intValue = longToNumber(reader.int64());
                    continue;
                case 4:
                    if (tag !== 33) {
                        break;
                    }
                    message.doubleValue = reader.double();
                    continue;
                case 5:
                    if (tag !== 42) {
                        break;
                    }
                    message.arrayValue = ArrayValue.decode(reader, reader.uint32());
                    continue;
                case 6:
                    if (tag !== 50) {
                        break;
                    }
                    message.kvlistValue = KeyValueList.decode(reader, reader.uint32());
                    continue;
                case 7:
                    if (tag !== 58) {
                        break;
                    }
                    message.bytesValue = reader.bytes();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            stringValue: isSet(object.stringValue)
                ? globalThis.String(object.stringValue)
                : undefined,
            boolValue: isSet(object.boolValue)
                ? globalThis.Boolean(object.boolValue)
                : undefined,
            intValue: isSet(object.intValue)
                ? globalThis.Number(object.intValue)
                : undefined,
            doubleValue: isSet(object.doubleValue)
                ? globalThis.Number(object.doubleValue)
                : undefined,
            arrayValue: isSet(object.arrayValue)
                ? ArrayValue.fromJSON(object.arrayValue)
                : undefined,
            kvlistValue: isSet(object.kvlistValue)
                ? KeyValueList.fromJSON(object.kvlistValue)
                : undefined,
            bytesValue: isSet(object.bytesValue)
                ? bytesFromBase64(object.bytesValue)
                : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.stringValue !== undefined) {
            obj.stringValue = message.stringValue;
        }
        if (message.boolValue !== undefined) {
            obj.boolValue = message.boolValue;
        }
        if (message.intValue !== undefined) {
            obj.intValue = Math.round(message.intValue);
        }
        if (message.doubleValue !== undefined) {
            obj.doubleValue = message.doubleValue;
        }
        if (message.arrayValue !== undefined) {
            obj.arrayValue = ArrayValue.toJSON(message.arrayValue);
        }
        if (message.kvlistValue !== undefined) {
            obj.kvlistValue = KeyValueList.toJSON(message.kvlistValue);
        }
        if (message.bytesValue !== undefined) {
            obj.bytesValue = base64FromBytes(message.bytesValue);
        }
        return obj;
    },
    create(base) {
        return AnyValue.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b, _c, _d, _e;
        const message = createBaseAnyValue();
        message.stringValue = (_a = object.stringValue) !== null && _a !== void 0 ? _a : undefined;
        message.boolValue = (_b = object.boolValue) !== null && _b !== void 0 ? _b : undefined;
        message.intValue = (_c = object.intValue) !== null && _c !== void 0 ? _c : undefined;
        message.doubleValue = (_d = object.doubleValue) !== null && _d !== void 0 ? _d : undefined;
        message.arrayValue =
            object.arrayValue !== undefined && object.arrayValue !== null
                ? ArrayValue.fromPartial(object.arrayValue)
                : undefined;
        message.kvlistValue =
            object.kvlistValue !== undefined && object.kvlistValue !== null
                ? KeyValueList.fromPartial(object.kvlistValue)
                : undefined;
        message.bytesValue = (_e = object.bytesValue) !== null && _e !== void 0 ? _e : undefined;
        return message;
    },
};
function createBaseArrayValue() {
    return { values: [] };
}
export const ArrayValue = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.values !== undefined && message.values.length !== 0) {
            for (const v of message.values) {
                AnyValue.encode(v, writer.uint32(10).fork()).ldelim();
            }
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseArrayValue();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.values.push(AnyValue.decode(reader, reader.uint32()));
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            values: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.values)
                ? object.values.map((e) => AnyValue.fromJSON(e))
                : [],
        };
    },
    toJSON(message) {
        var _a;
        const obj = {};
        if ((_a = message.values) === null || _a === void 0 ? void 0 : _a.length) {
            obj.values = message.values.map((e) => AnyValue.toJSON(e));
        }
        return obj;
    },
    create(base) {
        return ArrayValue.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a;
        const message = createBaseArrayValue();
        message.values = ((_a = object.values) === null || _a === void 0 ? void 0 : _a.map((e) => AnyValue.fromPartial(e))) || [];
        return message;
    },
};
function createBaseKeyValueList() {
    return { values: [] };
}
export const KeyValueList = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.values !== undefined && message.values.length !== 0) {
            for (const v of message.values) {
                KeyValue.encode(v, writer.uint32(10).fork()).ldelim();
            }
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseKeyValueList();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.values.push(KeyValue.decode(reader, reader.uint32()));
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            values: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.values)
                ? object.values.map((e) => KeyValue.fromJSON(e))
                : [],
        };
    },
    toJSON(message) {
        var _a;
        const obj = {};
        if ((_a = message.values) === null || _a === void 0 ? void 0 : _a.length) {
            obj.values = message.values.map((e) => KeyValue.toJSON(e));
        }
        return obj;
    },
    create(base) {
        return KeyValueList.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a;
        const message = createBaseKeyValueList();
        message.values = ((_a = object.values) === null || _a === void 0 ? void 0 : _a.map((e) => KeyValue.fromPartial(e))) || [];
        return message;
    },
};
function createBaseKeyValue() {
    return { key: '', value: undefined };
}
export const KeyValue = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.key !== undefined && message.key !== '') {
            writer.uint32(10).string(message.key);
        }
        if (message.value !== undefined) {
            AnyValue.encode(message.value, writer.uint32(18).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseKeyValue();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.key = reader.string();
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.value = AnyValue.decode(reader, reader.uint32());
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            key: isSet(object.key) ? globalThis.String(object.key) : '',
            value: isSet(object.value) ? AnyValue.fromJSON(object.value) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.key !== undefined && message.key !== '') {
            obj.key = message.key;
        }
        if (message.value !== undefined) {
            obj.value = AnyValue.toJSON(message.value);
        }
        return obj;
    },
    create(base) {
        return KeyValue.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a;
        const message = createBaseKeyValue();
        message.key = (_a = object.key) !== null && _a !== void 0 ? _a : '';
        message.value =
            object.value !== undefined && object.value !== null
                ? AnyValue.fromPartial(object.value)
                : undefined;
        return message;
    },
};
function createBaseInstrumentationLibrary() {
    return { name: '', version: '' };
}
export const InstrumentationLibrary = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.name !== undefined && message.name !== '') {
            writer.uint32(10).string(message.name);
        }
        if (message.version !== undefined && message.version !== '') {
            writer.uint32(18).string(message.version);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseInstrumentationLibrary();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.name = reader.string();
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.version = reader.string();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            name: isSet(object.name) ? globalThis.String(object.name) : '',
            version: isSet(object.version) ? globalThis.String(object.version) : '',
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.name !== undefined && message.name !== '') {
            obj.name = message.name;
        }
        if (message.version !== undefined && message.version !== '') {
            obj.version = message.version;
        }
        return obj;
    },
    create(base) {
        return InstrumentationLibrary.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b;
        const message = createBaseInstrumentationLibrary();
        message.name = (_a = object.name) !== null && _a !== void 0 ? _a : '';
        message.version = (_b = object.version) !== null && _b !== void 0 ? _b : '';
        return message;
    },
};
function createBaseInstrumentationScope() {
    return { name: '', version: '' };
}
export const InstrumentationScope = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.name !== undefined && message.name !== '') {
            writer.uint32(10).string(message.name);
        }
        if (message.version !== undefined && message.version !== '') {
            writer.uint32(18).string(message.version);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseInstrumentationScope();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.name = reader.string();
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.version = reader.string();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            name: isSet(object.name) ? globalThis.String(object.name) : '',
            version: isSet(object.version) ? globalThis.String(object.version) : '',
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.name !== undefined && message.name !== '') {
            obj.name = message.name;
        }
        if (message.version !== undefined && message.version !== '') {
            obj.version = message.version;
        }
        return obj;
    },
    create(base) {
        return InstrumentationScope.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b;
        const message = createBaseInstrumentationScope();
        message.name = (_a = object.name) !== null && _a !== void 0 ? _a : '';
        message.version = (_b = object.version) !== null && _b !== void 0 ? _b : '';
        return message;
    },
};
function bytesFromBase64(b64) {
    if (globalThis.Buffer) {
        return Uint8Array.from(globalThis.Buffer.from(b64, 'base64'));
    }
    else {
        const bin = globalThis.atob(b64);
        const arr = new Uint8Array(bin.length);
        for (let i = 0; i < bin.length; ++i) {
            arr[i] = bin.charCodeAt(i);
        }
        return arr;
    }
}
function base64FromBytes(arr) {
    if (globalThis.Buffer) {
        return globalThis.Buffer.from(arr).toString('base64');
    }
    else {
        const bin = [];
        arr.forEach((byte) => {
            bin.push(globalThis.String.fromCharCode(byte));
        });
        return globalThis.btoa(bin.join(''));
    }
}
function longToNumber(long) {
    if (long.gt(globalThis.Number.MAX_SAFE_INTEGER)) {
        throw new globalThis.Error('Value is larger than Number.MAX_SAFE_INTEGER');
    }
    return long.toNumber();
}
if (_m0.util.Long !== Long) {
    _m0.util.Long = Long;
    _m0.configure();
}
function isSet(value) {
    return value !== null && value !== undefined;
}
