// @hyperdx/otel-web/dist/esm/index.js
/*
Copyright 2020 Splunk Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
import { recordException as _recordException } from '@hyperdx/instrumentation-exception/build/src/browser';
import { ERROR_INSTRUMENTATION_NAME, HyperDXErrorInstrumentation, } from '@hyperdx/instrumentation-exception/build/src/browser/instrumentations/HyperDXErrorInstrumentation';
import { registerInstrumentations, } from '@opentelemetry/instrumentation';
import { ConsoleSpanExporter, SimpleSpanProcessor, AlwaysOffSampler, AlwaysOnSampler, ParentBasedSampler, } from '@opentelemetry/sdk-trace-base';
import { DiagLogLevel, diag, } from '@opentelemetry/api';
import { SplunkDocumentLoadInstrumentation } from './SplunkDocumentLoadInstrumentation';
import { SplunkUserInteractionInstrumentation, DEFAULT_AUTO_INSTRUMENTED_EVENTS, DEFAULT_AUTO_INSTRUMENTED_EVENT_NAMES, } from './SplunkUserInteractionInstrumentation';
import { OTLPTraceExporter } from '@opentelemetry/exporter-trace-otlp-http';
import { generateId, getPluginConfig } from './utils';
import { getRumSessionId, initSessionTracking } from './session';
import { SplunkWebSocketInstrumentation } from './SplunkWebSocketInstrumentation';
import { initWebVitals } from './webvitals';
import { SplunkLongTaskInstrumentation } from './SplunkLongTaskInstrumentation';
import { SplunkPageVisibilityInstrumentation } from './SplunkPageVisibilityInstrumentation';
import { SplunkConnectivityInstrumentation } from './SplunkConnectivityInstrumentation';
import { SplunkPostDocLoadResourceInstrumentation, } from './SplunkPostDocLoadResourceInstrumentation';
import { SplunkWebTracerProvider } from './SplunkWebTracerProvider';
import { InternalEventTarget } from './EventTarget';
import { SplunkContextManager, } from './SplunkContextManager';
import { Resource } from '@opentelemetry/resources';
import { SemanticResourceAttributes } from '@opentelemetry/semantic-conventions';
import { SDK_INFO, _globalThis } from '@opentelemetry/core';
import { VERSION } from './version';
import { getSyntheticsRunId, SYNTHETICS_RUN_ID_ATTRIBUTE } from './synthetics';
import { SplunkSpanAttributesProcessor } from './SplunkSpanAttributesProcessor';
import { SessionBasedSampler } from './SessionBasedSampler';
import { SplunkSocketIoClientInstrumentation, } from './SplunkSocketIoClientInstrumentation';
import { HyperDXBatchSpanProcessor } from './HyperDXBatchSpanProcessor';
import { HyperDXConsoleInstrumentation } from './HyperDXConsoleInstrumentation';
import { HyperDXFetchInstrumentation } from './HyperDXFetchInstrumentation';
import { HyperDXXMLHttpRequestInstrumentation } from './HyperDXXMLHttpRequestInstrumentation';
export * from './SessionBasedSampler';
export * from './SplunkWebTracerProvider';
const OPTIONS_DEFAULTS = {
    app: 'unknown-browser-app',
    url: undefined,
    bufferTimeout: 4000, //millis, tradeoff between batching and loss of spans by not sending before page close
    bufferSize: 50, // spans, tradeoff between batching and hitting sendBeacon invididual limits
    instrumentations: {},
    exporter: {
        factory: (options) => new OTLPTraceExporter({
            url: options.url,
            headers: {
                authorization: options.authHeader,
            },
        }),
    },
    spanProcessor: {
        factory: (exporter, config) => new HyperDXBatchSpanProcessor(exporter, config),
    },
    apiKey: undefined,
};
function migrateConfigOption(config, from, to) {
    if (from in config &&
        !(to in config && config[to] !== OPTIONS_DEFAULTS[to])) {
        // @ts-expect-error There's no way to type this right
        config[to] = config[from];
    }
}
/**
 * Update configuration based on configuration option renames
 */
function migrateConfig(config) {
    migrateConfigOption(config, 'app', 'applicationName');
    // migrateConfigOption(config, 'beaconUrl', 'beaconEndpoint');
    migrateConfigOption(config, 'environment', 'deploymentEnvironment');
    // migrateConfigOption(config, 'rumAuth', 'rumAccessToken');
    return config;
}
const INSTRUMENTATIONS = [
    {
        Instrument: SplunkDocumentLoadInstrumentation,
        confKey: 'document',
        disable: false,
    },
    {
        Instrument: HyperDXXMLHttpRequestInstrumentation,
        confKey: 'xhr',
        disable: false,
    },
    {
        Instrument: SplunkUserInteractionInstrumentation,
        confKey: 'interactions',
        disable: false,
    },
    {
        Instrument: SplunkPostDocLoadResourceInstrumentation,
        confKey: 'postload',
        disable: false,
    },
    { Instrument: HyperDXFetchInstrumentation, confKey: 'fetch', disable: false },
    {
        Instrument: SplunkWebSocketInstrumentation,
        confKey: 'websocket',
        disable: true,
    },
    {
        Instrument: SplunkLongTaskInstrumentation,
        confKey: 'longtask',
        disable: false,
    },
    {
        Instrument: HyperDXErrorInstrumentation,
        confKey: ERROR_INSTRUMENTATION_NAME,
        disable: false,
    },
    {
        Instrument: SplunkPageVisibilityInstrumentation,
        confKey: 'visibility',
        disable: true,
    },
    {
        Instrument: SplunkConnectivityInstrumentation,
        confKey: 'connectivity',
        disable: true,
    },
    {
        Instrument: SplunkSocketIoClientInstrumentation,
        confKey: 'socketio',
        disable: true,
    },
    {
        Instrument: HyperDXConsoleInstrumentation,
        confKey: 'console',
        disable: true,
    },
];
export const INSTRUMENTATIONS_ALL_DISABLED = INSTRUMENTATIONS.map((instrumentation) => instrumentation.confKey).reduce((acc, key) => {
    acc[key] = false;
    return acc;
}, { webvitals: false });
function buildExporter(options) {
    return options.exporter.factory({
        url: options.url,
        onAttributesSerializing: options.exporter.onAttributesSerializing,
        authHeader: options.apiKey,
    });
}
let inited = false;
let _deregisterInstrumentations;
let _deinitSessionTracking;
let _errorInstrumentation;
let _postDocLoadInstrumentation;
let eventTarget;
export const Rum = {
    DEFAULT_AUTO_INSTRUMENTED_EVENTS,
    DEFAULT_AUTO_INSTRUMENTED_EVENT_NAMES,
    // Re-export samplers as properties for easier use in CDN build
    AlwaysOnSampler,
    AlwaysOffSampler,
    ParentBasedSampler,
    SessionBasedSampler,
    get inited() {
        return inited;
    },
    _internalInit: function (options) {
        Rum.init(Object.assign(Object.assign({}, OPTIONS_DEFAULTS), options));
    },
    init: function (options) {
        // "env" based config still a bad idea for web
        if (!('OTEL_TRACES_EXPORTER' in _globalThis)) {
            _globalThis.OTEL_TRACES_EXPORTER = 'none';
        }
        const isOriginalConsole = (fn, name) => {
            // Sometimes __wrapped isn't defined on overwritten method, so we check if the fn name matches as a backup
            // I've only seen this occur in HMR environments
            return (typeof fn === 'function' && fn.__wrapped !== true && fn.name === name);
        };
        const consoleLog = isOriginalConsole(console === null || console === void 0 ? void 0 : console.log, 'log')
            ? console.log
            : () => { };
        const consoleWarn = isOriginalConsole(console === null || console === void 0 ? void 0 : console.warn, 'warn')
            ? console.warn
            : consoleLog;
        const consoleError = isOriginalConsole(console === null || console === void 0 ? void 0 : console.error, 'error')
            ? console.error
            : consoleLog;
        const consoleDebug = isOriginalConsole(console === null || console === void 0 ? void 0 : console.debug, 'debug')
            ? console.debug
            : consoleLog;
        const consoleTrace = isOriginalConsole(console === null || console === void 0 ? void 0 : console.trace, 'trace')
            ? console.trace
            : consoleLog;
        diag.setLogger({
            error: (...args) => {
                if (isOriginalConsole(consoleError, 'error')) {
                    consoleError(...args);
                }
            },
            warn: (...args) => {
                if (isOriginalConsole(consoleWarn, 'warn')) {
                    consoleWarn(...args);
                }
            },
            info: (...args) => {
                if (isOriginalConsole(consoleLog, 'log')) {
                    consoleLog(...args);
                }
            },
            debug: (...args) => {
                if (isOriginalConsole(consoleDebug, 'debug')) {
                    consoleDebug(...args);
                }
            },
            verbose: (...args) => {
                if (isOriginalConsole(consoleTrace, 'trace')) {
                    consoleTrace(...args);
                }
            },
        }, {
            suppressOverrideMessage: true,
            logLevel: (options === null || options === void 0 ? void 0 : options.debug) ? DiagLogLevel.DEBUG : DiagLogLevel.NONE,
        });
        if (typeof window !== 'object') {
            diag.error('Rum: Non-browser environment detected, aborting');
            return;
        }
        if (typeof Symbol !== 'function') {
            diag.error('Rum: browser not supported, disabling instrumentation.');
            return;
        }
        eventTarget = new InternalEventTarget();
        const processedOptions = Object.assign({}, OPTIONS_DEFAULTS, migrateConfig(options), {
            exporter: Object.assign({}, OPTIONS_DEFAULTS.exporter, options.exporter),
        });
        if (inited) {
            diag.warn('Rum already init()ed.');
            return;
        }
        if (!processedOptions.debug) {
            if (!processedOptions.url) {
                // eslint-disable-next-line quotes
                throw new Error("Rum.init( {url: 'https://something'} ) is required.");
            }
            else if (!processedOptions.url.startsWith('https') &&
                !processedOptions.allowInsecureUrl) {
                throw new Error('Not using https is unsafe, if you want to force it use allowInsecureUrl option.');
            }
        }
        const instanceId = generateId(64);
        _deinitSessionTracking = initSessionTracking(instanceId, eventTarget, processedOptions.cookieDomain).deinit;
        const { ignoreUrls, applicationName, deploymentEnvironment, version } = processedOptions;
        // enabled: false prevents registerInstrumentations from enabling instrumentations in constructor
        // they will be enabled in registerInstrumentations
        const pluginDefaults = { ignoreUrls, enabled: false };
        const resourceAttrs = Object.assign(Object.assign({}, SDK_INFO), { [SemanticResourceAttributes.TELEMETRY_SDK_NAME]: '@hyperdx/otel-web', [SemanticResourceAttributes.TELEMETRY_SDK_VERSION]: VERSION, [SemanticResourceAttributes.SERVICE_NAME]: applicationName, 
            // Splunk specific attributes
            'rum.version': VERSION, 'rum.scriptInstance': instanceId });
        const syntheticsRunId = getSyntheticsRunId();
        if (syntheticsRunId) {
            resourceAttrs[SYNTHETICS_RUN_ID_ATTRIBUTE] = syntheticsRunId;
        }
        const provider = new SplunkWebTracerProvider(Object.assign(Object.assign({}, processedOptions.tracer), { resource: new Resource(resourceAttrs) }));
        Object.defineProperty(provider.resource.attributes, 'rum.sessionId', {
            get() {
                return getRumSessionId();
            },
            configurable: true,
            enumerable: true,
        });
        const instrumentations = INSTRUMENTATIONS.map(({ Instrument, confKey, disable }) => {
            const pluginConf = getPluginConfig(processedOptions.instrumentations[confKey], pluginDefaults, disable);
            if (pluginConf) {
                // @ts-expect-error Can't mark in any way that processedOptions.instrumentations[confKey] is of specifc config type
                const instrumentation = new Instrument(pluginConf);
                if (confKey === ERROR_INSTRUMENTATION_NAME &&
                    instrumentation instanceof HyperDXErrorInstrumentation) {
                    _errorInstrumentation = instrumentation;
                }
                if (confKey === 'postload' &&
                    instrumentation instanceof SplunkPostDocLoadResourceInstrumentation) {
                    _postDocLoadInstrumentation = instrumentation;
                }
                return instrumentation;
            }
            return null;
        }).filter((a) => Boolean(a));
        this.attributesProcessor = new SplunkSpanAttributesProcessor(Object.assign(Object.assign(Object.assign({}, (deploymentEnvironment
            ? {
                environment: deploymentEnvironment,
                'deployment.environment': deploymentEnvironment,
            }
            : {})), (version ? { 'app.version': version } : {})), (processedOptions.globalAttributes || {})));
        provider.addSpanProcessor(this.attributesProcessor);
        if (processedOptions.url) {
            const exporter = buildExporter(processedOptions);
            const spanProcessor = processedOptions.spanProcessor.factory(exporter, {
                scheduledDelayMillis: processedOptions.bufferTimeout,
                maxExportBatchSize: processedOptions.bufferSize,
            });
            provider.addSpanProcessor(spanProcessor);
            this._processor = spanProcessor;
        }
        if (processedOptions.debug) {
            provider.addSpanProcessor(new SimpleSpanProcessor(new ConsoleSpanExporter()));
        }
        window.addEventListener('visibilitychange', () => {
            // this condition applies when the page is hidden or when it's closed
            // see for more details: https://developers.google.com/web/updates/2018/07/page-lifecycle-api#developer-recommendations-for-each-state
            if (document.visibilityState === 'hidden') {
                this._processor.forceFlush();
            }
        });
        provider.register({
            contextManager: new SplunkContextManager(Object.assign(Object.assign({}, processedOptions.context), { onBeforeContextStart: () => _postDocLoadInstrumentation === null || _postDocLoadInstrumentation === void 0 ? void 0 : _postDocLoadInstrumentation.onBeforeContextChange(), onBeforeContextEnd: () => _postDocLoadInstrumentation === null || _postDocLoadInstrumentation === void 0 ? void 0 : _postDocLoadInstrumentation.onBeforeContextChange() })),
        });
        // After context manager registration so instrumentation event listeners are affected accordingly
        _deregisterInstrumentations = registerInstrumentations({
            tracerProvider: provider,
            instrumentations,
        });
        this.provider = provider;
        const vitalsConf = getPluginConfig(processedOptions.instrumentations.webvitals);
        if (vitalsConf !== false) {
            initWebVitals(provider);
        }
        inited = true;
        diag.info('Rum.init() complete');
    },
    deinit() {
        if (!inited) {
            return;
        }
        _deregisterInstrumentations === null || _deregisterInstrumentations === void 0 ? void 0 : _deregisterInstrumentations();
        _deregisterInstrumentations = undefined;
        _deinitSessionTracking === null || _deinitSessionTracking === void 0 ? void 0 : _deinitSessionTracking();
        _deinitSessionTracking = undefined;
        this.provider.shutdown();
        delete this.provider;
        eventTarget = undefined;
        diag.disable();
        inited = false;
    },
    setGlobalAttributes(attributes) {
        var _a, _b;
        (_a = this.attributesProcessor) === null || _a === void 0 ? void 0 : _a.setGlobalAttributes(attributes);
        eventTarget === null || eventTarget === void 0 ? void 0 : eventTarget.emit('global-attributes-changed', {
            attributes: ((_b = this.attributesProcessor) === null || _b === void 0 ? void 0 : _b.getGlobalAttributes()) || {},
        });
    },
    getGlobalAttributes() {
        var _a;
        return ((_a = this.attributesProcessor) === null || _a === void 0 ? void 0 : _a.getGlobalAttributes()) || {};
    },
    // add custom action
    addAction(name, attributes) {
        if (!inited) {
            diag.debug('Rum not inited');
            return;
        }
        const now = Date.now();
        const tracer = this.provider.getTracer('custom-action');
        const span = tracer.startSpan(name, { startTime: now });
        span.setAttributes(attributes);
        span.end(now);
    },
    _experimental_getGlobalAttributes() {
        return this.getGlobalAttributes();
    },
    recordException(error, attributes) {
        if (!inited) {
            diag.debug('Rum not inited');
            return;
        }
        const tracer = this.provider.getTracer('record-exception');
        return _recordException(error, {
            tracer,
            attributes,
        });
    },
    error(...args) {
        if (!inited) {
            diag.debug('Rum not inited');
            return;
        }
        if (!_errorInstrumentation) {
            diag.error('Error was reported, but error instrumentation is disabled.');
            return;
        }
        _errorInstrumentation.hdxReport('Rum.error', args);
    },
    addEventListener(name, callback) {
        eventTarget === null || eventTarget === void 0 ? void 0 : eventTarget.addEventListener(name, callback);
    },
    removeEventListener(name, callback) {
        eventTarget === null || eventTarget === void 0 ? void 0 : eventTarget.removeEventListener(name, callback);
    },
    _experimental_addEventListener(name, callback) {
        return this.addEventListener(name, callback);
    },
    _experimental_removeEventListener(name, callback) {
        return this.removeEventListener(name, callback);
    },
    getSessionId() {
        return getRumSessionId();
    },
    _experimental_getSessionId() {
        return this.getSessionId();
    },
};
export default Rum;
