// @hyperdx/instrumentation-exception/build/src/instrumentation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ExceptionInstrumentation = void 0;
const instrumentation_1 = require("@opentelemetry/instrumentation");
const api_1 = require("@opentelemetry/api");
const package_json_1 = require("../package.json");
const onuncaughtexception_1 = require("./node/integrations/onuncaughtexception");
const onunhandledrejection_1 = require("./node/integrations/onunhandledrejection");
class ExceptionInstrumentation extends instrumentation_1.InstrumentationBase {
    constructor(config = {}) {
        super(package_json_1.name, package_json_1.version, config);
    }
    setConfig(config = {}) {
        this._config = Object.assign({}, config);
    }
    getConfig() {
        return this._config;
    }
    init() {
    }
    setTracerProvider(tracerProvider) {
        super.setTracerProvider(tracerProvider);
        this._traceForceFlusher = this._traceForceFlush(tracerProvider);
    }
    _traceForceFlush(tracerProvider) {
        if (!tracerProvider)
            return undefined;
        let currentProvider = tracerProvider;
        if (typeof currentProvider.getDelegate === 'function') {
            currentProvider = currentProvider.getDelegate();
        }
        if (typeof currentProvider.forceFlush === 'function') {
            return currentProvider.forceFlush.bind(currentProvider);
        }
        return undefined;
    }
    async forceFlush() {
        const flushers = [];
        if (this._traceForceFlusher) {
            flushers.push(this._traceForceFlusher());
        }
        else {
            api_1.diag.error('Spans may not be exported because we are not force flushing before callback.');
        }
        await Promise.all(flushers);
    }
    enable() {
        var _a;
        const config = this.getConfig();
        const ff = (_a = config._internalForceFlush) !== null && _a !== void 0 ? _a : this.forceFlush;
        (0, onuncaughtexception_1.onUncaughtExceptionIntegration)({
            exitEvenIfOtherHandlersAreRegistered: config.exitEvenIfOtherHandlersAreRegistered,
            forceFlush: () => ff(),
        }).setup({});
        (0, onunhandledrejection_1.onUnhandledRejectionIntegration)({
            mode: config.unhandledRejectionMode,
            forceFlush: () => ff(),
        }).setup({});
    }
    disable() {
    }
}
exports.ExceptionInstrumentation = ExceptionInstrumentation;
