// @hyperdx/instrumentation-exception/build/src/node/integrations/local-variables/local-variables-sync.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.localVariablesSyncIntegration = exports.createCallbackList = void 0;
const core_1 = require("@sentry/core");
const utils_1 = require("@sentry/utils");
const nodeVersion_1 = require("../../nodeVersion");
const common_1 = require("./common");
const api_1 = require("../../sdk/api");
function createCallbackList(complete) {
    let callbacks = [];
    let completedCalled = false;
    function checkedComplete(result) {
        callbacks = [];
        if (completedCalled) {
            return;
        }
        completedCalled = true;
        complete(result);
    }
    callbacks.push(checkedComplete);
    function add(fn) {
        callbacks.push(fn);
    }
    function next(result) {
        const popped = callbacks.pop() || checkedComplete;
        try {
            popped(result);
        }
        catch (_) {
            checkedComplete(result);
        }
    }
    return { add, next };
}
exports.createCallbackList = createCallbackList;
class AsyncSession {
    constructor(_session) {
        this._session = _session;
    }
    static async create(orDefault) {
        if (orDefault) {
            return orDefault;
        }
        const inspector = await Promise.resolve().then(() => __importStar(require('node:inspector')));
        return new AsyncSession(new inspector.Session());
    }
    configureAndConnect(onPause, captureAll) {
        this._session.connect();
        this._session.on('Debugger.paused', (event) => {
            onPause(event, () => {
                this._session.post('Debugger.resume');
            });
        });
        this._session.post('Debugger.enable');
        this._session.post('Debugger.setPauseOnExceptions', {
            state: captureAll ? 'all' : 'uncaught',
        });
    }
    setPauseOnExceptions(captureAll) {
        this._session.post('Debugger.setPauseOnExceptions', {
            state: captureAll ? 'all' : 'uncaught',
        });
    }
    getLocalVariables(objectId, complete) {
        this._getProperties(objectId, (props) => {
            var _a, _b, _c;
            const { add, next } = createCallbackList(complete);
            for (const prop of props) {
                if (((_a = prop === null || prop === void 0 ? void 0 : prop.value) === null || _a === void 0 ? void 0 : _a.objectId) && (prop === null || prop === void 0 ? void 0 : prop.value.className) === 'Array') {
                    const id = prop.value.objectId;
                    add((vars) => this._unrollArray(id, prop.name, vars, next));
                }
                else if (((_b = prop === null || prop === void 0 ? void 0 : prop.value) === null || _b === void 0 ? void 0 : _b.objectId) &&
                    ((_c = prop === null || prop === void 0 ? void 0 : prop.value) === null || _c === void 0 ? void 0 : _c.className) === 'Object') {
                    const id = prop.value.objectId;
                    add((vars) => this._unrollObject(id, prop.name, vars, next));
                }
                else if (prop === null || prop === void 0 ? void 0 : prop.value) {
                    add((vars) => this._unrollOther(prop, vars, next));
                }
            }
            next({});
        });
    }
    _getProperties(objectId, next) {
        this._session.post('Runtime.getProperties', {
            objectId,
            ownProperties: true,
        }, (err, params) => {
            if (err) {
                next([]);
            }
            else {
                next(params.result);
            }
        });
    }
    _unrollArray(objectId, name, vars, next) {
        this._getProperties(objectId, (props) => {
            vars[name] = props
                .filter((v) => v.name !== 'length' && !isNaN(parseInt(v.name, 10)))
                .sort((a, b) => parseInt(a.name, 10) - parseInt(b.name, 10))
                .map((v) => { var _a; return (_a = v === null || v === void 0 ? void 0 : v.value) === null || _a === void 0 ? void 0 : _a.value; });
            next(vars);
        });
    }
    _unrollObject(objectId, name, vars, next) {
        this._getProperties(objectId, (props) => {
            vars[name] = props
                .map((v) => { var _a; return [v.name, (_a = v === null || v === void 0 ? void 0 : v.value) === null || _a === void 0 ? void 0 : _a.value]; })
                .reduce((obj, [key, val]) => {
                obj[key] = val;
                return obj;
            }, {});
            next(vars);
        });
    }
    _unrollOther(prop, vars, next) {
        if (prop.value) {
            if ('value' in prop.value) {
                if (prop.value.value === undefined || prop.value.value === null) {
                    vars[prop.name] = `<${prop.value.value}>`;
                }
                else {
                    vars[prop.name] = prop.value.value;
                }
            }
            else if ('description' in prop.value &&
                prop.value.type !== 'function') {
                vars[prop.name] = `<${prop.value.description}>`;
            }
            else if (prop.value.type === 'undefined') {
                vars[prop.name] = '<undefined>';
            }
        }
        next(vars);
    }
}
const INTEGRATION_NAME = 'LocalVariables';
const _localVariablesSyncIntegration = ((options = {}, sessionOverride) => {
    const cachedFrames = new utils_1.LRUMap(20);
    let rateLimiter;
    let shouldProcessEvent = false;
    function addLocalVariablesToException(exception) {
        var _a, _b;
        const hash = (0, common_1.hashFrames)((_a = exception === null || exception === void 0 ? void 0 : exception.stacktrace) === null || _a === void 0 ? void 0 : _a.frames);
        if (hash === undefined) {
            return;
        }
        const cachedFrame = cachedFrames.remove(hash);
        if (cachedFrame === undefined) {
            return;
        }
        const frames = (((_b = exception.stacktrace) === null || _b === void 0 ? void 0 : _b.frames) || []).filter((frame) => frame.function !== 'new Promise');
        for (let i = 0; i < frames.length; i++) {
            const frameIndex = frames.length - i - 1;
            if (!frames[frameIndex] || !cachedFrame[i]) {
                break;
            }
            if (cachedFrame[i].vars === undefined ||
                frames[frameIndex].in_app === false ||
                !(0, common_1.functionNamesMatch)(frames[frameIndex].function, cachedFrame[i].function)) {
                continue;
            }
            frames[frameIndex].vars = cachedFrame[i].vars;
        }
    }
    function addLocalVariablesToEvent(event) {
        var _a;
        for (const exception of ((_a = event === null || event === void 0 ? void 0 : event.exception) === null || _a === void 0 ? void 0 : _a.values) || []) {
            addLocalVariablesToException(exception);
        }
        return event;
    }
    return {
        name: INTEGRATION_NAME,
        setupOnce() {
            const unsupportedNodeVersion = nodeVersion_1.NODE_MAJOR < 18;
            if (unsupportedNodeVersion) {
                utils_1.logger.log('The `LocalVariables` integration is only supported on Node >= v18.');
                return;
            }
            AsyncSession.create(sessionOverride).then((session) => {
                function handlePaused(stackParser, { params: { reason, data, callFrames }, }, complete) {
                    if (reason !== 'exception' && reason !== 'promiseRejection') {
                        complete();
                        return;
                    }
                    rateLimiter === null || rateLimiter === void 0 ? void 0 : rateLimiter();
                    const exceptionHash = (0, common_1.hashFromStack)(stackParser, data === null || data === void 0 ? void 0 : data.description);
                    if (exceptionHash == undefined) {
                        complete();
                        return;
                    }
                    const { add, next } = createCallbackList((frames) => {
                        cachedFrames.set(exceptionHash, frames);
                        complete();
                    });
                    for (let i = 0; i < Math.min(callFrames.length, 5); i++) {
                        const { scopeChain, functionName, this: obj } = callFrames[i];
                        const localScope = scopeChain.find((scope) => scope.type === 'local');
                        const fn = obj.className === 'global' || !obj.className
                            ? functionName
                            : `${obj.className}.${functionName}`;
                        if ((localScope === null || localScope === void 0 ? void 0 : localScope.object.objectId) === undefined) {
                            add((frames) => {
                                frames[i] = { function: fn };
                                next(frames);
                            });
                        }
                        else {
                            const id = localScope.object.objectId;
                            add((frames) => session === null || session === void 0 ? void 0 : session.getLocalVariables(id, (vars) => {
                                frames[i] = { function: fn, vars };
                                next(frames);
                            }));
                        }
                    }
                    next([]);
                }
                const captureAll = options.captureAllExceptions !== false;
                session.configureAndConnect((ev, complete) => handlePaused(api_1.defaultStackParser, ev, complete), captureAll);
                if (captureAll) {
                    const max = options.maxExceptionsPerSecond || 50;
                    rateLimiter = (0, common_1.createRateLimiter)(max, () => {
                        utils_1.logger.log('Local variables rate-limit lifted.');
                        session === null || session === void 0 ? void 0 : session.setPauseOnExceptions(true);
                    }, (seconds) => {
                        utils_1.logger.log(`Local variables rate-limit exceeded. Disabling capturing of caught exceptions for ${seconds} seconds.`);
                        session === null || session === void 0 ? void 0 : session.setPauseOnExceptions(false);
                    });
                }
                shouldProcessEvent = true;
            }, (error) => {
                utils_1.logger.log('The `LocalVariables` integration failed to start.', error);
            });
        },
        processEvent(event) {
            if (shouldProcessEvent) {
                return addLocalVariablesToEvent(event);
            }
            return event;
        },
        _getCachedFramesCount() {
            return cachedFrames.size;
        },
        _getFirstCachedFrame() {
            return cachedFrames.values()[0];
        },
    };
});
exports.localVariablesSyncIntegration = (0, core_1.defineIntegration)(_localVariablesSyncIntegration);
