// @hyperdx/instrumentation-exception/build/src/node/integrations/local-variables/local-variables-async.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.localVariablesAsyncIntegration = exports.base64WorkerScript = void 0;
const node_worker_threads_1 = require("node:worker_threads");
const core_1 = require("@sentry/core");
const utils_1 = require("@sentry/utils");
const common_1 = require("./common");
exports.base64WorkerScript = '###LocalVariablesWorkerScript###';
function log(...args) {
    utils_1.logger.log('[LocalVariables]', ...args);
}
exports.localVariablesAsyncIntegration = (0, core_1.defineIntegration)(((integrationOptions = {}) => {
    const cachedFrames = new utils_1.LRUMap(20);
    function addLocalVariablesToException(exception) {
        var _a, _b;
        const hash = (0, common_1.hashFrames)((_a = exception === null || exception === void 0 ? void 0 : exception.stacktrace) === null || _a === void 0 ? void 0 : _a.frames);
        if (hash === undefined) {
            return;
        }
        const cachedFrame = cachedFrames.remove(hash);
        if (cachedFrame === undefined) {
            return;
        }
        const frames = (((_b = exception.stacktrace) === null || _b === void 0 ? void 0 : _b.frames) || []).filter((frame) => frame.function !== 'new Promise');
        for (let i = 0; i < frames.length; i++) {
            const frameIndex = frames.length - i - 1;
            if (!frames[frameIndex] || !cachedFrame[i]) {
                break;
            }
            if (cachedFrame[i].vars === undefined ||
                frames[frameIndex].in_app === false ||
                !(0, common_1.functionNamesMatch)(frames[frameIndex].function, cachedFrame[i].function)) {
                continue;
            }
            frames[frameIndex].vars = cachedFrame[i].vars;
        }
    }
    function addLocalVariablesToEvent(event) {
        var _a;
        for (const exception of ((_a = event.exception) === null || _a === void 0 ? void 0 : _a.values) || []) {
            addLocalVariablesToException(exception);
        }
        return event;
    }
    async function startInspector() {
        const inspector = await Promise.resolve().then(() => __importStar(require('node:inspector')));
        if (!inspector.url()) {
            inspector.open(0);
        }
    }
    function startWorker(options) {
        const worker = new node_worker_threads_1.Worker(new URL(`data:application/javascript;base64,${exports.base64WorkerScript}`), {
            workerData: options,
            execArgv: [],
        });
        process.on('exit', () => {
            worker.terminate();
        });
        worker.on('message', ({ exceptionHash, frames }) => {
            cachedFrames.set(exceptionHash, frames);
        });
        worker.once('error', (err) => {
            log('Worker error', err);
        });
        worker.once('exit', (code) => {
            log('Worker exit', code);
        });
        worker.unref();
    }
    return {
        name: 'LocalVariablesAsync',
        setup() {
            const options = {
                ...integrationOptions,
                debug: utils_1.logger.isEnabled(),
            };
            startInspector().then(() => {
                try {
                    startWorker(options);
                }
                catch (e) {
                    utils_1.logger.error('Failed to start worker', e);
                }
            }, (e) => {
                utils_1.logger.error('Failed to start inspector', e);
            });
        },
        processEvent(event) {
            return addLocalVariablesToEvent(event);
        },
    };
}));
