// @hyperdx/instrumentation-exception/build/src/node/integrations/local-variables/worker.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const promises_1 = require("node:inspector/promises");
const node_worker_threads_1 = require("node:worker_threads");
const utils_1 = require("@sentry/utils");
const module_1 = require("../../utils/module");
const common_1 = require("./common");
const options = node_worker_threads_1.workerData;
const stackParser = (0, utils_1.createStackParser)((0, utils_1.nodeStackLineParser)((0, module_1.createGetModuleFromFilename)(options.basePath)));
function log(...args) {
    if (options.debug) {
        console.log('[LocalVariables Worker]', ...args);
    }
}
async function unrollArray(session, objectId, name, vars) {
    const properties = await session.post('Runtime.getProperties', {
        objectId,
        ownProperties: true,
    });
    vars[name] = properties.result
        .filter((v) => v.name !== 'length' && !isNaN(parseInt(v.name, 10)))
        .sort((a, b) => parseInt(a.name, 10) - parseInt(b.name, 10))
        .map((v) => { var _a; return (_a = v.value) === null || _a === void 0 ? void 0 : _a.value; });
}
async function unrollObject(session, objectId, name, vars) {
    const properties = await session.post('Runtime.getProperties', {
        objectId,
        ownProperties: true,
    });
    vars[name] = properties.result
        .map((v) => { var _a; return [v.name, (_a = v.value) === null || _a === void 0 ? void 0 : _a.value]; })
        .reduce((obj, [key, val]) => {
        obj[key] = val;
        return obj;
    }, {});
}
function unrollOther(prop, vars) {
    if (!prop.value) {
        return;
    }
    if ('value' in prop.value) {
        if (prop.value.value === undefined || prop.value.value === null) {
            vars[prop.name] = `<${prop.value.value}>`;
        }
        else {
            vars[prop.name] = prop.value.value;
        }
    }
    else if ('description' in prop.value && prop.value.type !== 'function') {
        vars[prop.name] = `<${prop.value.description}>`;
    }
    else if (prop.value.type === 'undefined') {
        vars[prop.name] = '<undefined>';
    }
}
async function getLocalVariables(session, objectId) {
    var _a, _b, _c;
    const properties = await session.post('Runtime.getProperties', {
        objectId,
        ownProperties: true,
    });
    const variables = {};
    for (const prop of properties.result) {
        if (((_a = prop === null || prop === void 0 ? void 0 : prop.value) === null || _a === void 0 ? void 0 : _a.objectId) && (prop === null || prop === void 0 ? void 0 : prop.value.className) === 'Array') {
            const id = prop.value.objectId;
            await unrollArray(session, id, prop.name, variables);
        }
        else if (((_b = prop === null || prop === void 0 ? void 0 : prop.value) === null || _b === void 0 ? void 0 : _b.objectId) && ((_c = prop === null || prop === void 0 ? void 0 : prop.value) === null || _c === void 0 ? void 0 : _c.className) === 'Object') {
            const id = prop.value.objectId;
            await unrollObject(session, id, prop.name, variables);
        }
        else if (prop === null || prop === void 0 ? void 0 : prop.value) {
            unrollOther(prop, variables);
        }
    }
    return variables;
}
let rateLimiter;
async function handlePaused(session, stackParser, { reason, data, callFrames }) {
    if (reason !== 'exception' && reason !== 'promiseRejection') {
        return;
    }
    rateLimiter === null || rateLimiter === void 0 ? void 0 : rateLimiter();
    const exceptionHash = (0, common_1.hashFromStack)(stackParser, data === null || data === void 0 ? void 0 : data.description);
    if (exceptionHash == undefined) {
        return;
    }
    const frames = [];
    for (let i = 0; i < callFrames.length; i++) {
        const { scopeChain, functionName, this: obj } = callFrames[i];
        const localScope = scopeChain.find((scope) => scope.type === 'local');
        const fn = obj.className === 'global' || !obj.className
            ? functionName
            : `${obj.className}.${functionName}`;
        if ((localScope === null || localScope === void 0 ? void 0 : localScope.object.objectId) === undefined) {
            frames[i] = { function: fn };
        }
        else {
            const vars = await getLocalVariables(session, localScope.object.objectId);
            frames[i] = { function: fn, vars };
        }
    }
    node_worker_threads_1.parentPort === null || node_worker_threads_1.parentPort === void 0 ? void 0 : node_worker_threads_1.parentPort.postMessage({ exceptionHash, frames });
}
async function startDebugger() {
    const session = new promises_1.Session();
    session.connectToMainThread();
    log('Connected to main thread');
    let isPaused = false;
    session.on('Debugger.resumed', () => {
        isPaused = false;
    });
    session.on('Debugger.paused', (event) => {
        isPaused = true;
        handlePaused(session, stackParser, event.params).then(() => {
            return isPaused ? session.post('Debugger.resume') : Promise.resolve();
        }, (_) => {
        });
    });
    await session.post('Debugger.enable');
    const captureAll = options.captureAllExceptions !== false;
    await session.post('Debugger.setPauseOnExceptions', {
        state: captureAll ? 'all' : 'uncaught',
    });
    if (captureAll) {
        const max = options.maxExceptionsPerSecond || 50;
        rateLimiter = (0, common_1.createRateLimiter)(max, async () => {
            log('Rate-limit lifted.');
            await session.post('Debugger.setPauseOnExceptions', { state: 'all' });
        }, async (seconds) => {
            log(`Rate-limit exceeded. Disabling capturing of caught exceptions for ${seconds} seconds.`);
            await session.post('Debugger.setPauseOnExceptions', {
                state: 'uncaught',
            });
        });
    }
}
startDebugger().catch((e) => {
    log('Failed to start debugger', e);
});
setInterval(() => {
}, 10000);
