// @hyperdx/instrumentation-exception/build/src/browser/instrumentations/HyperDXErrorInstrumentation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HyperDXErrorInstrumentation = exports.ERROR_INSTRUMENTATION_VERSION = exports.ERROR_INSTRUMENTATION_NAME = void 0;
const tslib_1 = require("tslib");
const shimmer = tslib_1.__importStar(require("shimmer"));
const instrumentation_1 = require("@opentelemetry/instrumentation");
const __1 = require("../");
const utils_1 = require("./utils");
const STACK_LIMIT = 4096;
const MESSAGE_LIMIT = 1024;
function useful(s) {
    return s && s.trim() !== '' && !s.startsWith('[object') && s !== 'error';
}
function stringifyValue(value) {
    if (value === undefined) {
        return '(undefined)';
    }
    return value.toString();
}
function addStackIfUseful(span, err) {
    if (err && err.stack && useful(err.stack)) {
        span.setAttribute('error.stack', (0, utils_1.limitLen)(err.stack.toString(), STACK_LIMIT));
    }
}
exports.ERROR_INSTRUMENTATION_NAME = 'errors';
exports.ERROR_INSTRUMENTATION_VERSION = '1';
class HyperDXErrorInstrumentation extends instrumentation_1.InstrumentationBase {
    constructor(config) {
        super(exports.ERROR_INSTRUMENTATION_NAME, exports.ERROR_INSTRUMENTATION_VERSION, config);
        this._consoleErrorHandler = (original) => {
            return (...args) => {
                this.hdxReport('console.error', args);
                return original.apply(this, args);
            };
        };
        this._unhandledRejectionListener = (event) => {
            this.hdxReport('unhandledrejection', event.reason);
        };
        this._errorListener = (event) => {
            this.hdxReport('onerror', event);
        };
        this._documentErrorListener = (event) => {
            this.hdxReport('eventListener.error', event);
        };
    }
    init() { }
    enable() {
        shimmer.wrap(console, 'error', this._consoleErrorHandler);
        window.addEventListener('unhandledrejection', this._unhandledRejectionListener);
        window.addEventListener('error', this._errorListener);
        document.documentElement.addEventListener('error', this._documentErrorListener, { capture: true });
    }
    disable() {
        shimmer.unwrap(console, 'error');
        window.removeEventListener('unhandledrejection', this._unhandledRejectionListener);
        window.removeEventListener('error', this._errorListener);
        document.documentElement.removeEventListener('error', this._documentErrorListener, { capture: true });
    }
    hdxReportError(source, err) {
        const msg = err.message || err.toString();
        if (!useful(msg) && !err.stack) {
            return;
        }
        const now = Date.now();
        const span = this.tracer.startSpan(source, { startTime: now });
        span.setAttribute('component', 'error');
        span.setAttribute('error', true);
        span.setAttribute('error.object', useful(err.name)
            ? err.name
            : err.constructor && err.constructor.name
                ? err.constructor.name
                : 'Error');
        span.setAttribute('error.message', (0, utils_1.limitLen)(msg, MESSAGE_LIMIT));
        addStackIfUseful(span, err);
        (0, __1.recordException)(err, {
            tracer: this.tracer,
            span,
        }).finally(() => {
            span.end(now);
        });
    }
    hdxReportString(source, message, firstError) {
        if (!useful(message)) {
            return;
        }
        const now = Date.now();
        const span = this.tracer.startSpan(source, { startTime: now });
        span.setAttribute('component', 'error');
        span.setAttribute('error', true);
        span.setAttribute('error.object', 'String');
        span.setAttribute('error.message', (0, utils_1.limitLen)(message, MESSAGE_LIMIT));
        if (firstError) {
            addStackIfUseful(span, firstError);
            (0, __1.recordException)(firstError, {
                tracer: this.tracer,
                span,
            }).finally(() => {
                span.end(now);
            });
        }
        else {
            span.end(now);
        }
    }
    hdxReportErrorEvent(source, ev) {
        if (ev.error) {
            this.hdxReport(source, ev.error);
        }
        else if (ev.message) {
            this.hdxReport(source, ev.message);
        }
    }
    hdxReportEvent(source, ev) {
        if (!ev.target && !useful(ev.type)) {
            return;
        }
        const now = Date.now();
        const span = this.tracer.startSpan(source, { startTime: now });
        span.setAttribute('component', 'error');
        span.setAttribute('error.type', ev.type);
        if (ev.target) {
            span.setAttribute('target_element', ev.target.tagName);
            span.setAttribute('target_xpath', (0, utils_1.getElementXPath)(ev.target, true));
            span.setAttribute('target_src', ev.target.src);
        }
        (0, __1.recordException)(ev, {
            tracer: this.tracer,
            span,
        }).finally(() => {
            span.end(now);
        });
    }
    hdxReport(source, arg) {
        if (Array.isArray(arg) && arg.length === 0) {
            return;
        }
        if (arg instanceof Array && arg.length === 1) {
            arg = arg[0];
        }
        if (arg instanceof Error) {
            this.hdxReportError(source, arg);
        }
        else if (arg instanceof ErrorEvent) {
            this.hdxReportErrorEvent(source, arg);
        }
        else if (arg instanceof Event) {
            this.hdxReportEvent(source, arg);
        }
        else if (typeof arg === 'string') {
            this.hdxReportString(source, arg);
        }
        else if (arg instanceof Array) {
            const firstError = arg.find((x) => x instanceof Error);
            this.hdxReportString(source, arg.map((x) => stringifyValue(x)).join(' '), firstError);
        }
        else {
            this.hdxReportString(source, stringifyValue(arg));
        }
    }
}
exports.HyperDXErrorInstrumentation = HyperDXErrorInstrumentation;
