// @hyperdx/otel-web/dist/cjs/index.js
"use strict";
/*
Copyright 2020 Splunk Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Rum = exports.INSTRUMENTATIONS_ALL_DISABLED = void 0;
const browser_1 = require("@hyperdx/instrumentation-exception/build/src/browser");
const HyperDXErrorInstrumentation_1 = require("@hyperdx/instrumentation-exception/build/src/browser/instrumentations/HyperDXErrorInstrumentation");
const instrumentation_1 = require("@opentelemetry/instrumentation");
const sdk_trace_base_1 = require("@opentelemetry/sdk-trace-base");
const api_1 = require("@opentelemetry/api");
const SplunkDocumentLoadInstrumentation_1 = require("./SplunkDocumentLoadInstrumentation");
const SplunkUserInteractionInstrumentation_1 = require("./SplunkUserInteractionInstrumentation");
const exporter_trace_otlp_http_1 = require("@opentelemetry/exporter-trace-otlp-http");
const utils_1 = require("./utils");
const session_1 = require("./session");
const SplunkWebSocketInstrumentation_1 = require("./SplunkWebSocketInstrumentation");
const webvitals_1 = require("./webvitals");
const SplunkLongTaskInstrumentation_1 = require("./SplunkLongTaskInstrumentation");
const SplunkPageVisibilityInstrumentation_1 = require("./SplunkPageVisibilityInstrumentation");
const SplunkConnectivityInstrumentation_1 = require("./SplunkConnectivityInstrumentation");
const SplunkPostDocLoadResourceInstrumentation_1 = require("./SplunkPostDocLoadResourceInstrumentation");
const SplunkWebTracerProvider_1 = require("./SplunkWebTracerProvider");
const EventTarget_1 = require("./EventTarget");
const SplunkContextManager_1 = require("./SplunkContextManager");
const resources_1 = require("@opentelemetry/resources");
const semantic_conventions_1 = require("@opentelemetry/semantic-conventions");
const core_1 = require("@opentelemetry/core");
const version_1 = require("./version");
const synthetics_1 = require("./synthetics");
const SplunkSpanAttributesProcessor_1 = require("./SplunkSpanAttributesProcessor");
const SessionBasedSampler_1 = require("./SessionBasedSampler");
const SplunkSocketIoClientInstrumentation_1 = require("./SplunkSocketIoClientInstrumentation");
const HyperDXBatchSpanProcessor_1 = require("./HyperDXBatchSpanProcessor");
const HyperDXConsoleInstrumentation_1 = require("./HyperDXConsoleInstrumentation");
const HyperDXFetchInstrumentation_1 = require("./HyperDXFetchInstrumentation");
const HyperDXXMLHttpRequestInstrumentation_1 = require("./HyperDXXMLHttpRequestInstrumentation");
__exportStar(require("./SessionBasedSampler"), exports);
__exportStar(require("./SplunkWebTracerProvider"), exports);
const OPTIONS_DEFAULTS = {
    app: 'unknown-browser-app',
    url: undefined,
    bufferTimeout: 4000, //millis, tradeoff between batching and loss of spans by not sending before page close
    bufferSize: 50, // spans, tradeoff between batching and hitting sendBeacon invididual limits
    instrumentations: {},
    exporter: {
        factory: (options) => new exporter_trace_otlp_http_1.OTLPTraceExporter({
            url: options.url,
            headers: {
                authorization: options.authHeader,
            },
        }),
    },
    spanProcessor: {
        factory: (exporter, config) => new HyperDXBatchSpanProcessor_1.HyperDXBatchSpanProcessor(exporter, config),
    },
    apiKey: undefined,
};
function migrateConfigOption(config, from, to) {
    if (from in config &&
        !(to in config && config[to] !== OPTIONS_DEFAULTS[to])) {
        // @ts-expect-error There's no way to type this right
        config[to] = config[from];
    }
}
/**
 * Update configuration based on configuration option renames
 */
function migrateConfig(config) {
    migrateConfigOption(config, 'app', 'applicationName');
    // migrateConfigOption(config, 'beaconUrl', 'beaconEndpoint');
    migrateConfigOption(config, 'environment', 'deploymentEnvironment');
    // migrateConfigOption(config, 'rumAuth', 'rumAccessToken');
    return config;
}
const INSTRUMENTATIONS = [
    {
        Instrument: SplunkDocumentLoadInstrumentation_1.SplunkDocumentLoadInstrumentation,
        confKey: 'document',
        disable: false,
    },
    {
        Instrument: HyperDXXMLHttpRequestInstrumentation_1.HyperDXXMLHttpRequestInstrumentation,
        confKey: 'xhr',
        disable: false,
    },
    {
        Instrument: SplunkUserInteractionInstrumentation_1.SplunkUserInteractionInstrumentation,
        confKey: 'interactions',
        disable: false,
    },
    {
        Instrument: SplunkPostDocLoadResourceInstrumentation_1.SplunkPostDocLoadResourceInstrumentation,
        confKey: 'postload',
        disable: false,
    },
    { Instrument: HyperDXFetchInstrumentation_1.HyperDXFetchInstrumentation, confKey: 'fetch', disable: false },
    {
        Instrument: SplunkWebSocketInstrumentation_1.SplunkWebSocketInstrumentation,
        confKey: 'websocket',
        disable: true,
    },
    {
        Instrument: SplunkLongTaskInstrumentation_1.SplunkLongTaskInstrumentation,
        confKey: 'longtask',
        disable: false,
    },
    {
        Instrument: HyperDXErrorInstrumentation_1.HyperDXErrorInstrumentation,
        confKey: HyperDXErrorInstrumentation_1.ERROR_INSTRUMENTATION_NAME,
        disable: false,
    },
    {
        Instrument: SplunkPageVisibilityInstrumentation_1.SplunkPageVisibilityInstrumentation,
        confKey: 'visibility',
        disable: true,
    },
    {
        Instrument: SplunkConnectivityInstrumentation_1.SplunkConnectivityInstrumentation,
        confKey: 'connectivity',
        disable: true,
    },
    {
        Instrument: SplunkSocketIoClientInstrumentation_1.SplunkSocketIoClientInstrumentation,
        confKey: 'socketio',
        disable: true,
    },
    {
        Instrument: HyperDXConsoleInstrumentation_1.HyperDXConsoleInstrumentation,
        confKey: 'console',
        disable: true,
    },
];
exports.INSTRUMENTATIONS_ALL_DISABLED = INSTRUMENTATIONS.map((instrumentation) => instrumentation.confKey).reduce((acc, key) => {
    acc[key] = false;
    return acc;
}, { webvitals: false });
function buildExporter(options) {
    return options.exporter.factory({
        url: options.url,
        onAttributesSerializing: options.exporter.onAttributesSerializing,
        authHeader: options.apiKey,
    });
}
let inited = false;
let _deregisterInstrumentations;
let _deinitSessionTracking;
let _errorInstrumentation;
let _postDocLoadInstrumentation;
let eventTarget;
exports.Rum = {
    DEFAULT_AUTO_INSTRUMENTED_EVENTS: SplunkUserInteractionInstrumentation_1.DEFAULT_AUTO_INSTRUMENTED_EVENTS,
    DEFAULT_AUTO_INSTRUMENTED_EVENT_NAMES: SplunkUserInteractionInstrumentation_1.DEFAULT_AUTO_INSTRUMENTED_EVENT_NAMES,
    // Re-export samplers as properties for easier use in CDN build
    AlwaysOnSampler: sdk_trace_base_1.AlwaysOnSampler,
    AlwaysOffSampler: sdk_trace_base_1.AlwaysOffSampler,
    ParentBasedSampler: sdk_trace_base_1.ParentBasedSampler,
    SessionBasedSampler: SessionBasedSampler_1.SessionBasedSampler,
    get inited() {
        return inited;
    },
    _internalInit: function (options) {
        exports.Rum.init(Object.assign(Object.assign({}, OPTIONS_DEFAULTS), options));
    },
    init: function (options) {
        // "env" based config still a bad idea for web
        if (!('OTEL_TRACES_EXPORTER' in core_1._globalThis)) {
            core_1._globalThis.OTEL_TRACES_EXPORTER = 'none';
        }
        const isOriginalConsole = (fn, name) => {
            // Sometimes __wrapped isn't defined on overwritten method, so we check if the fn name matches as a backup
            // I've only seen this occur in HMR environments
            return (typeof fn === 'function' && fn.__wrapped !== true && fn.name === name);
        };
        const consoleLog = isOriginalConsole(console === null || console === void 0 ? void 0 : console.log, 'log')
            ? console.log
            : () => { };
        const consoleWarn = isOriginalConsole(console === null || console === void 0 ? void 0 : console.warn, 'warn')
            ? console.warn
            : consoleLog;
        const consoleError = isOriginalConsole(console === null || console === void 0 ? void 0 : console.error, 'error')
            ? console.error
            : consoleLog;
        const consoleDebug = isOriginalConsole(console === null || console === void 0 ? void 0 : console.debug, 'debug')
            ? console.debug
            : consoleLog;
        const consoleTrace = isOriginalConsole(console === null || console === void 0 ? void 0 : console.trace, 'trace')
            ? console.trace
            : consoleLog;
        api_1.diag.setLogger({
            error: (...args) => {
                if (isOriginalConsole(consoleError, 'error')) {
                    consoleError(...args);
                }
            },
            warn: (...args) => {
                if (isOriginalConsole(consoleWarn, 'warn')) {
                    consoleWarn(...args);
                }
            },
            info: (...args) => {
                if (isOriginalConsole(consoleLog, 'log')) {
                    consoleLog(...args);
                }
            },
            debug: (...args) => {
                if (isOriginalConsole(consoleDebug, 'debug')) {
                    consoleDebug(...args);
                }
            },
            verbose: (...args) => {
                if (isOriginalConsole(consoleTrace, 'trace')) {
                    consoleTrace(...args);
                }
            },
        }, {
            suppressOverrideMessage: true,
            logLevel: (options === null || options === void 0 ? void 0 : options.debug) ? api_1.DiagLogLevel.DEBUG : api_1.DiagLogLevel.NONE,
        });
        if (typeof window !== 'object') {
            api_1.diag.error('Rum: Non-browser environment detected, aborting');
            return;
        }
        if (typeof Symbol !== 'function') {
            api_1.diag.error('Rum: browser not supported, disabling instrumentation.');
            return;
        }
        eventTarget = new EventTarget_1.InternalEventTarget();
        const processedOptions = Object.assign({}, OPTIONS_DEFAULTS, migrateConfig(options), {
            exporter: Object.assign({}, OPTIONS_DEFAULTS.exporter, options.exporter),
        });
        if (inited) {
            api_1.diag.warn('Rum already init()ed.');
            return;
        }
        if (!processedOptions.debug) {
            if (!processedOptions.url) {
                // eslint-disable-next-line quotes
                throw new Error("Rum.init( {url: 'https://something'} ) is required.");
            }
            else if (!processedOptions.url.startsWith('https') &&
                !processedOptions.allowInsecureUrl) {
                throw new Error('Not using https is unsafe, if you want to force it use allowInsecureUrl option.');
            }
        }
        const instanceId = (0, utils_1.generateId)(64);
        _deinitSessionTracking = (0, session_1.initSessionTracking)(instanceId, eventTarget, processedOptions.cookieDomain).deinit;
        const { ignoreUrls, applicationName, deploymentEnvironment, version } = processedOptions;
        // enabled: false prevents registerInstrumentations from enabling instrumentations in constructor
        // they will be enabled in registerInstrumentations
        const pluginDefaults = { ignoreUrls, enabled: false };
        const resourceAttrs = Object.assign(Object.assign({}, core_1.SDK_INFO), { [semantic_conventions_1.SemanticResourceAttributes.TELEMETRY_SDK_NAME]: '@hyperdx/otel-web', [semantic_conventions_1.SemanticResourceAttributes.TELEMETRY_SDK_VERSION]: version_1.VERSION, [semantic_conventions_1.SemanticResourceAttributes.SERVICE_NAME]: applicationName, 
            // Splunk specific attributes
            'rum.version': version_1.VERSION, 'rum.scriptInstance': instanceId });
        const syntheticsRunId = (0, synthetics_1.getSyntheticsRunId)();
        if (syntheticsRunId) {
            resourceAttrs[synthetics_1.SYNTHETICS_RUN_ID_ATTRIBUTE] = syntheticsRunId;
        }
        const provider = new SplunkWebTracerProvider_1.SplunkWebTracerProvider(Object.assign(Object.assign({}, processedOptions.tracer), { resource: new resources_1.Resource(resourceAttrs) }));
        Object.defineProperty(provider.resource.attributes, 'rum.sessionId', {
            get() {
                return (0, session_1.getRumSessionId)();
            },
            configurable: true,
            enumerable: true,
        });
        const instrumentations = INSTRUMENTATIONS.map(({ Instrument, confKey, disable }) => {
            const pluginConf = (0, utils_1.getPluginConfig)(processedOptions.instrumentations[confKey], pluginDefaults, disable);
            if (pluginConf) {
                // @ts-expect-error Can't mark in any way that processedOptions.instrumentations[confKey] is of specifc config type
                const instrumentation = new Instrument(pluginConf);
                if (confKey === HyperDXErrorInstrumentation_1.ERROR_INSTRUMENTATION_NAME &&
                    instrumentation instanceof HyperDXErrorInstrumentation_1.HyperDXErrorInstrumentation) {
                    _errorInstrumentation = instrumentation;
                }
                if (confKey === 'postload' &&
                    instrumentation instanceof SplunkPostDocLoadResourceInstrumentation_1.SplunkPostDocLoadResourceInstrumentation) {
                    _postDocLoadInstrumentation = instrumentation;
                }
                return instrumentation;
            }
            return null;
        }).filter((a) => Boolean(a));
        this.attributesProcessor = new SplunkSpanAttributesProcessor_1.SplunkSpanAttributesProcessor(Object.assign(Object.assign(Object.assign({}, (deploymentEnvironment
            ? {
                environment: deploymentEnvironment,
                'deployment.environment': deploymentEnvironment,
            }
            : {})), (version ? { 'app.version': version } : {})), (processedOptions.globalAttributes || {})));
        provider.addSpanProcessor(this.attributesProcessor);
        if (processedOptions.url) {
            const exporter = buildExporter(processedOptions);
            const spanProcessor = processedOptions.spanProcessor.factory(exporter, {
                scheduledDelayMillis: processedOptions.bufferTimeout,
                maxExportBatchSize: processedOptions.bufferSize,
            });
            provider.addSpanProcessor(spanProcessor);
            this._processor = spanProcessor;
        }
        if (processedOptions.debug) {
            provider.addSpanProcessor(new sdk_trace_base_1.SimpleSpanProcessor(new sdk_trace_base_1.ConsoleSpanExporter()));
        }
        window.addEventListener('visibilitychange', () => {
            // this condition applies when the page is hidden or when it's closed
            // see for more details: https://developers.google.com/web/updates/2018/07/page-lifecycle-api#developer-recommendations-for-each-state
            if (document.visibilityState === 'hidden') {
                this._processor.forceFlush();
            }
        });
        provider.register({
            contextManager: new SplunkContextManager_1.SplunkContextManager(Object.assign(Object.assign({}, processedOptions.context), { onBeforeContextStart: () => _postDocLoadInstrumentation === null || _postDocLoadInstrumentation === void 0 ? void 0 : _postDocLoadInstrumentation.onBeforeContextChange(), onBeforeContextEnd: () => _postDocLoadInstrumentation === null || _postDocLoadInstrumentation === void 0 ? void 0 : _postDocLoadInstrumentation.onBeforeContextChange() })),
        });
        // After context manager registration so instrumentation event listeners are affected accordingly
        _deregisterInstrumentations = (0, instrumentation_1.registerInstrumentations)({
            tracerProvider: provider,
            instrumentations,
        });
        this.provider = provider;
        const vitalsConf = (0, utils_1.getPluginConfig)(processedOptions.instrumentations.webvitals);
        if (vitalsConf !== false) {
            (0, webvitals_1.initWebVitals)(provider);
        }
        inited = true;
        api_1.diag.info('Rum.init() complete');
    },
    deinit() {
        if (!inited) {
            return;
        }
        _deregisterInstrumentations === null || _deregisterInstrumentations === void 0 ? void 0 : _deregisterInstrumentations();
        _deregisterInstrumentations = undefined;
        _deinitSessionTracking === null || _deinitSessionTracking === void 0 ? void 0 : _deinitSessionTracking();
        _deinitSessionTracking = undefined;
        this.provider.shutdown();
        delete this.provider;
        eventTarget = undefined;
        api_1.diag.disable();
        inited = false;
    },
    setGlobalAttributes(attributes) {
        var _a, _b;
        (_a = this.attributesProcessor) === null || _a === void 0 ? void 0 : _a.setGlobalAttributes(attributes);
        eventTarget === null || eventTarget === void 0 ? void 0 : eventTarget.emit('global-attributes-changed', {
            attributes: ((_b = this.attributesProcessor) === null || _b === void 0 ? void 0 : _b.getGlobalAttributes()) || {},
        });
    },
    getGlobalAttributes() {
        var _a;
        return ((_a = this.attributesProcessor) === null || _a === void 0 ? void 0 : _a.getGlobalAttributes()) || {};
    },
    // add custom action
    addAction(name, attributes) {
        if (!inited) {
            api_1.diag.debug('Rum not inited');
            return;
        }
        const now = Date.now();
        const tracer = this.provider.getTracer('custom-action');
        const span = tracer.startSpan(name, { startTime: now });
        span.setAttributes(attributes);
        span.end(now);
    },
    _experimental_getGlobalAttributes() {
        return this.getGlobalAttributes();
    },
    recordException(error, attributes) {
        if (!inited) {
            api_1.diag.debug('Rum not inited');
            return;
        }
        const tracer = this.provider.getTracer('record-exception');
        return (0, browser_1.recordException)(error, {
            tracer,
            attributes,
        });
    },
    error(...args) {
        if (!inited) {
            api_1.diag.debug('Rum not inited');
            return;
        }
        if (!_errorInstrumentation) {
            api_1.diag.error('Error was reported, but error instrumentation is disabled.');
            return;
        }
        _errorInstrumentation.hdxReport('Rum.error', args);
    },
    addEventListener(name, callback) {
        eventTarget === null || eventTarget === void 0 ? void 0 : eventTarget.addEventListener(name, callback);
    },
    removeEventListener(name, callback) {
        eventTarget === null || eventTarget === void 0 ? void 0 : eventTarget.removeEventListener(name, callback);
    },
    _experimental_addEventListener(name, callback) {
        return this.addEventListener(name, callback);
    },
    _experimental_removeEventListener(name, callback) {
        return this.removeEventListener(name, callback);
    },
    getSessionId() {
        return (0, session_1.getRumSessionId)();
    },
    _experimental_getSessionId() {
        return this.getSessionId();
    },
};
exports.default = exports.Rum;
