// @hyperdx/otel-web/dist/cjs/utils.js
"use strict";
/*
Copyright 2020 Splunk Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.headerCapture = exports.waitForGlobal = exports.getOriginalFunction = exports.wrapNatively = exports.isIframe = exports.isFunction = exports.getPluginConfig = exports.limitLen = exports.findCookieValue = exports.generateId = exports.jsonToString = void 0;
const json_stringify_safe_1 = require("json-stringify-safe");
const shimmer_1 = require("shimmer");
const jsonToString = (json) => {
    let output = '[HyperDX] Failed to stringify';
    let error = false;
    try {
        output = JSON.stringify(json);
    }
    catch (ex) {
        error = true;
        // ignore error
    }
    if (error) {
        try {
            output = (0, json_stringify_safe_1.default)(json);
        }
        catch (ex) {
            // ignore error
        }
    }
    return output;
};
exports.jsonToString = jsonToString;
function generateId(bits) {
    const xes = 'x'.repeat(bits / 4);
    return xes.replace(/x/g, function () {
        return ((Math.random() * 16) | 0).toString(16);
    });
}
exports.generateId = generateId;
function findCookieValue(cookieName) {
    const decodedCookie = decodeURIComponent(document.cookie);
    const cookies = decodedCookie.split(';');
    for (let i = 0; i < cookies.length; i++) {
        const c = cookies[i].trim();
        if (c.indexOf(cookieName + '=') === 0) {
            return c.substring((cookieName + '=').length, c.length);
        }
    }
    return undefined;
}
exports.findCookieValue = findCookieValue;
function limitLen(s, cap) {
    if (s.length > cap) {
        return s.substring(0, cap);
    }
    else {
        return s;
    }
}
exports.limitLen = limitLen;
/**
 * Get plugin config from user provided value
 *
 * @template {Object|undefined} T
 * @param value Value given by user
 * @param defaults Default value
 * @param defaultDisable If undefined by user should mean false
 */
function getPluginConfig(value, defaults, defaultDisable) {
    if (value === false) {
        return value;
    }
    if (value === undefined && defaultDisable) {
        return false;
    }
    return Object.assign({}, defaults, value);
}
exports.getPluginConfig = getPluginConfig;
/**
 * Type guard that checks if value is function (and acts as user-defined type guard)
 *
 * @param value Value to check
 * @returns is function
 */
function isFunction(value) {
    return typeof value === 'function';
}
exports.isFunction = isFunction;
function isIframe() {
    try {
        return window.self !== window.top;
    }
    catch (e) {
        return true;
    }
}
exports.isIframe = isIframe;
/**
 * Wrap function while keeping the toString calling the original as some frameworks
 * use it to determine if the function's native or polyfilled
 *
 * Example:
 * https://github.com/vuejs/vue/blob/0603ff695d2f41286239298210113cbe2b209e28/src/core/util/env.js#L58
 * https://github.com/vuejs/vue/blob/0603ff695d2f41286239298210113cbe2b209e28/src/core/util/next-tick.js#L42
 *
 * @param nodule Target object
 * @param name Property to patch
 * @param wrapper Wrapper
 */
// eslint-disable-next-line @typescript-eslint/ban-types
function wrapNatively(nodule, name, wrapper) {
    const orig = nodule[name];
    (0, shimmer_1.wrap)(nodule, name, wrapper);
    const wrapped = nodule[name];
    wrapped.toString = orig.toString.bind(orig);
}
exports.wrapNatively = wrapNatively;
/**
 * Get the original version of function (without all of the shimmer wrappings)
 */
function getOriginalFunction(func) {
    // @ts-expect-error __original isn't mentioned in types
    while (func.__original && func.__original !== func) {
        // @ts-expect-error same
        func = func.__original;
    }
    return func;
}
exports.getOriginalFunction = getOriginalFunction;
/**
 * Wait for a variable to be set globally
 *
 * @param {string} identifier Name of the variable (window[identifier])
 * @param {function} callback Fired when such value is available
 * @returns {function} cleanup to call in disable (in case not defined before instrumentation is disabled)
 */
function waitForGlobal(identifier, callback) {
    if (window[identifier]) {
        callback(window[identifier]);
        return () => { };
    }
    const value = window[identifier]; // most cases undefined
    let used = false;
    Object.defineProperty(window, identifier, {
        get() {
            return value;
        },
        set(newVal) {
            delete window[identifier];
            used = true;
            window[identifier] = newVal;
            callback(newVal);
        },
        configurable: true,
        enumerable: false,
    });
    return () => {
        // Don't touch if value is used or another defineProperty used it
        if (used || window[identifier] !== value) {
            return;
        }
        delete window[identifier];
        if (value !== undefined) {
            window[identifier] = value;
        }
    };
}
exports.waitForGlobal = waitForGlobal;
// https://github.com/open-telemetry/opentelemetry-js/blob/b400c2e5d9729c3528482781a93393602dc6dc9f/experimental/packages/opentelemetry-instrumentation-http/src/utils.ts#L573
function headerCapture(type, headers) {
    const normalizedHeaders = new Map(headers.map((header) => [header, header.toLowerCase().replace(/-/g, '_')]));
    return (span, getHeader) => {
        for (const [capturedHeader, normalizedHeader] of normalizedHeaders) {
            const value = getHeader(capturedHeader);
            if (value === undefined) {
                continue;
            }
            const key = `http.${type}.header.${normalizedHeader}`;
            if (typeof value === 'string') {
                span.setAttribute(key, [value]);
            }
            else if (Array.isArray(value)) {
                span.setAttribute(key, value);
            }
            else {
                span.setAttribute(key, [value]);
            }
        }
    };
}
exports.headerCapture = headerCapture;
