// @hyperdx/instrumentation-exception/build/src/node/integrations/tracing/express.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.setupExpressErrorHandler = exports.expressErrorHandler = void 0;
const __1 = require("../../");
function expressErrorHandler(options) {
    return function sentryErrorMiddleware(error, _req, res, next) {
        const shouldHandleError = (options === null || options === void 0 ? void 0 : options.shouldHandleError) || defaultShouldHandleError;
        if (shouldHandleError(error)) {
            (0, __1.recordException)(error, {
                mechanism: { type: 'middleware', handled: false },
            });
            next(error);
            return;
        }
        next(error);
    };
}
exports.expressErrorHandler = expressErrorHandler;
function setupExpressErrorHandler(app) {
    app.use(expressErrorHandler());
}
exports.setupExpressErrorHandler = setupExpressErrorHandler;
function getStatusCodeFromResponse(error) {
    const statusCode = error.status ||
        error.statusCode ||
        error.status_code ||
        (error.output && error.output.statusCode);
    return statusCode ? parseInt(statusCode, 10) : 500;
}
function defaultShouldHandleError(error) {
    const status = getStatusCodeFromResponse(error);
    return status >= 500;
}
