// @hyperdx/otel-web-session-recorder/dist/cjs/opentelemetry/proto/logs/v1/logs.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LogRecord = exports.InstrumentationLibraryLogs = exports.ScopeLogs = exports.ResourceLogs = exports.LogsData = exports.logRecordFlagsToJSON = exports.logRecordFlagsFromJSON = exports.LogRecordFlags = exports.severityNumberToJSON = exports.severityNumberFromJSON = exports.SeverityNumber = exports.protobufPackage = void 0;
/* eslint-disable */
const long_1 = require("long");
const minimal_1 = require("protobufjs/minimal");
const common_1 = require("../../common/v1/common");
const resource_1 = require("../../resource/v1/resource");
exports.protobufPackage = 'opentelemetry.proto.logs.v1';
/** Possible values for LogRecord.SeverityNumber. */
var SeverityNumber;
(function (SeverityNumber) {
    /** SEVERITY_NUMBER_UNSPECIFIED - UNSPECIFIED is the default SeverityNumber, it MUST NOT be used. */
    SeverityNumber[SeverityNumber["SEVERITY_NUMBER_UNSPECIFIED"] = 0] = "SEVERITY_NUMBER_UNSPECIFIED";
    SeverityNumber[SeverityNumber["SEVERITY_NUMBER_TRACE"] = 1] = "SEVERITY_NUMBER_TRACE";
    SeverityNumber[SeverityNumber["SEVERITY_NUMBER_TRACE2"] = 2] = "SEVERITY_NUMBER_TRACE2";
    SeverityNumber[SeverityNumber["SEVERITY_NUMBER_TRACE3"] = 3] = "SEVERITY_NUMBER_TRACE3";
    SeverityNumber[SeverityNumber["SEVERITY_NUMBER_TRACE4"] = 4] = "SEVERITY_NUMBER_TRACE4";
    SeverityNumber[SeverityNumber["SEVERITY_NUMBER_DEBUG"] = 5] = "SEVERITY_NUMBER_DEBUG";
    SeverityNumber[SeverityNumber["SEVERITY_NUMBER_DEBUG2"] = 6] = "SEVERITY_NUMBER_DEBUG2";
    SeverityNumber[SeverityNumber["SEVERITY_NUMBER_DEBUG3"] = 7] = "SEVERITY_NUMBER_DEBUG3";
    SeverityNumber[SeverityNumber["SEVERITY_NUMBER_DEBUG4"] = 8] = "SEVERITY_NUMBER_DEBUG4";
    SeverityNumber[SeverityNumber["SEVERITY_NUMBER_INFO"] = 9] = "SEVERITY_NUMBER_INFO";
    SeverityNumber[SeverityNumber["SEVERITY_NUMBER_INFO2"] = 10] = "SEVERITY_NUMBER_INFO2";
    SeverityNumber[SeverityNumber["SEVERITY_NUMBER_INFO3"] = 11] = "SEVERITY_NUMBER_INFO3";
    SeverityNumber[SeverityNumber["SEVERITY_NUMBER_INFO4"] = 12] = "SEVERITY_NUMBER_INFO4";
    SeverityNumber[SeverityNumber["SEVERITY_NUMBER_WARN"] = 13] = "SEVERITY_NUMBER_WARN";
    SeverityNumber[SeverityNumber["SEVERITY_NUMBER_WARN2"] = 14] = "SEVERITY_NUMBER_WARN2";
    SeverityNumber[SeverityNumber["SEVERITY_NUMBER_WARN3"] = 15] = "SEVERITY_NUMBER_WARN3";
    SeverityNumber[SeverityNumber["SEVERITY_NUMBER_WARN4"] = 16] = "SEVERITY_NUMBER_WARN4";
    SeverityNumber[SeverityNumber["SEVERITY_NUMBER_ERROR"] = 17] = "SEVERITY_NUMBER_ERROR";
    SeverityNumber[SeverityNumber["SEVERITY_NUMBER_ERROR2"] = 18] = "SEVERITY_NUMBER_ERROR2";
    SeverityNumber[SeverityNumber["SEVERITY_NUMBER_ERROR3"] = 19] = "SEVERITY_NUMBER_ERROR3";
    SeverityNumber[SeverityNumber["SEVERITY_NUMBER_ERROR4"] = 20] = "SEVERITY_NUMBER_ERROR4";
    SeverityNumber[SeverityNumber["SEVERITY_NUMBER_FATAL"] = 21] = "SEVERITY_NUMBER_FATAL";
    SeverityNumber[SeverityNumber["SEVERITY_NUMBER_FATAL2"] = 22] = "SEVERITY_NUMBER_FATAL2";
    SeverityNumber[SeverityNumber["SEVERITY_NUMBER_FATAL3"] = 23] = "SEVERITY_NUMBER_FATAL3";
    SeverityNumber[SeverityNumber["SEVERITY_NUMBER_FATAL4"] = 24] = "SEVERITY_NUMBER_FATAL4";
    SeverityNumber[SeverityNumber["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
})(SeverityNumber || (exports.SeverityNumber = SeverityNumber = {}));
function severityNumberFromJSON(object) {
    switch (object) {
        case 0:
        case 'SEVERITY_NUMBER_UNSPECIFIED':
            return SeverityNumber.SEVERITY_NUMBER_UNSPECIFIED;
        case 1:
        case 'SEVERITY_NUMBER_TRACE':
            return SeverityNumber.SEVERITY_NUMBER_TRACE;
        case 2:
        case 'SEVERITY_NUMBER_TRACE2':
            return SeverityNumber.SEVERITY_NUMBER_TRACE2;
        case 3:
        case 'SEVERITY_NUMBER_TRACE3':
            return SeverityNumber.SEVERITY_NUMBER_TRACE3;
        case 4:
        case 'SEVERITY_NUMBER_TRACE4':
            return SeverityNumber.SEVERITY_NUMBER_TRACE4;
        case 5:
        case 'SEVERITY_NUMBER_DEBUG':
            return SeverityNumber.SEVERITY_NUMBER_DEBUG;
        case 6:
        case 'SEVERITY_NUMBER_DEBUG2':
            return SeverityNumber.SEVERITY_NUMBER_DEBUG2;
        case 7:
        case 'SEVERITY_NUMBER_DEBUG3':
            return SeverityNumber.SEVERITY_NUMBER_DEBUG3;
        case 8:
        case 'SEVERITY_NUMBER_DEBUG4':
            return SeverityNumber.SEVERITY_NUMBER_DEBUG4;
        case 9:
        case 'SEVERITY_NUMBER_INFO':
            return SeverityNumber.SEVERITY_NUMBER_INFO;
        case 10:
        case 'SEVERITY_NUMBER_INFO2':
            return SeverityNumber.SEVERITY_NUMBER_INFO2;
        case 11:
        case 'SEVERITY_NUMBER_INFO3':
            return SeverityNumber.SEVERITY_NUMBER_INFO3;
        case 12:
        case 'SEVERITY_NUMBER_INFO4':
            return SeverityNumber.SEVERITY_NUMBER_INFO4;
        case 13:
        case 'SEVERITY_NUMBER_WARN':
            return SeverityNumber.SEVERITY_NUMBER_WARN;
        case 14:
        case 'SEVERITY_NUMBER_WARN2':
            return SeverityNumber.SEVERITY_NUMBER_WARN2;
        case 15:
        case 'SEVERITY_NUMBER_WARN3':
            return SeverityNumber.SEVERITY_NUMBER_WARN3;
        case 16:
        case 'SEVERITY_NUMBER_WARN4':
            return SeverityNumber.SEVERITY_NUMBER_WARN4;
        case 17:
        case 'SEVERITY_NUMBER_ERROR':
            return SeverityNumber.SEVERITY_NUMBER_ERROR;
        case 18:
        case 'SEVERITY_NUMBER_ERROR2':
            return SeverityNumber.SEVERITY_NUMBER_ERROR2;
        case 19:
        case 'SEVERITY_NUMBER_ERROR3':
            return SeverityNumber.SEVERITY_NUMBER_ERROR3;
        case 20:
        case 'SEVERITY_NUMBER_ERROR4':
            return SeverityNumber.SEVERITY_NUMBER_ERROR4;
        case 21:
        case 'SEVERITY_NUMBER_FATAL':
            return SeverityNumber.SEVERITY_NUMBER_FATAL;
        case 22:
        case 'SEVERITY_NUMBER_FATAL2':
            return SeverityNumber.SEVERITY_NUMBER_FATAL2;
        case 23:
        case 'SEVERITY_NUMBER_FATAL3':
            return SeverityNumber.SEVERITY_NUMBER_FATAL3;
        case 24:
        case 'SEVERITY_NUMBER_FATAL4':
            return SeverityNumber.SEVERITY_NUMBER_FATAL4;
        case -1:
        case 'UNRECOGNIZED':
        default:
            return SeverityNumber.UNRECOGNIZED;
    }
}
exports.severityNumberFromJSON = severityNumberFromJSON;
function severityNumberToJSON(object) {
    switch (object) {
        case SeverityNumber.SEVERITY_NUMBER_UNSPECIFIED:
            return 'SEVERITY_NUMBER_UNSPECIFIED';
        case SeverityNumber.SEVERITY_NUMBER_TRACE:
            return 'SEVERITY_NUMBER_TRACE';
        case SeverityNumber.SEVERITY_NUMBER_TRACE2:
            return 'SEVERITY_NUMBER_TRACE2';
        case SeverityNumber.SEVERITY_NUMBER_TRACE3:
            return 'SEVERITY_NUMBER_TRACE3';
        case SeverityNumber.SEVERITY_NUMBER_TRACE4:
            return 'SEVERITY_NUMBER_TRACE4';
        case SeverityNumber.SEVERITY_NUMBER_DEBUG:
            return 'SEVERITY_NUMBER_DEBUG';
        case SeverityNumber.SEVERITY_NUMBER_DEBUG2:
            return 'SEVERITY_NUMBER_DEBUG2';
        case SeverityNumber.SEVERITY_NUMBER_DEBUG3:
            return 'SEVERITY_NUMBER_DEBUG3';
        case SeverityNumber.SEVERITY_NUMBER_DEBUG4:
            return 'SEVERITY_NUMBER_DEBUG4';
        case SeverityNumber.SEVERITY_NUMBER_INFO:
            return 'SEVERITY_NUMBER_INFO';
        case SeverityNumber.SEVERITY_NUMBER_INFO2:
            return 'SEVERITY_NUMBER_INFO2';
        case SeverityNumber.SEVERITY_NUMBER_INFO3:
            return 'SEVERITY_NUMBER_INFO3';
        case SeverityNumber.SEVERITY_NUMBER_INFO4:
            return 'SEVERITY_NUMBER_INFO4';
        case SeverityNumber.SEVERITY_NUMBER_WARN:
            return 'SEVERITY_NUMBER_WARN';
        case SeverityNumber.SEVERITY_NUMBER_WARN2:
            return 'SEVERITY_NUMBER_WARN2';
        case SeverityNumber.SEVERITY_NUMBER_WARN3:
            return 'SEVERITY_NUMBER_WARN3';
        case SeverityNumber.SEVERITY_NUMBER_WARN4:
            return 'SEVERITY_NUMBER_WARN4';
        case SeverityNumber.SEVERITY_NUMBER_ERROR:
            return 'SEVERITY_NUMBER_ERROR';
        case SeverityNumber.SEVERITY_NUMBER_ERROR2:
            return 'SEVERITY_NUMBER_ERROR2';
        case SeverityNumber.SEVERITY_NUMBER_ERROR3:
            return 'SEVERITY_NUMBER_ERROR3';
        case SeverityNumber.SEVERITY_NUMBER_ERROR4:
            return 'SEVERITY_NUMBER_ERROR4';
        case SeverityNumber.SEVERITY_NUMBER_FATAL:
            return 'SEVERITY_NUMBER_FATAL';
        case SeverityNumber.SEVERITY_NUMBER_FATAL2:
            return 'SEVERITY_NUMBER_FATAL2';
        case SeverityNumber.SEVERITY_NUMBER_FATAL3:
            return 'SEVERITY_NUMBER_FATAL3';
        case SeverityNumber.SEVERITY_NUMBER_FATAL4:
            return 'SEVERITY_NUMBER_FATAL4';
        case SeverityNumber.UNRECOGNIZED:
        default:
            return 'UNRECOGNIZED';
    }
}
exports.severityNumberToJSON = severityNumberToJSON;
/** Masks for LogRecord.flags field. */
var LogRecordFlags;
(function (LogRecordFlags) {
    LogRecordFlags[LogRecordFlags["LOG_RECORD_FLAG_UNSPECIFIED"] = 0] = "LOG_RECORD_FLAG_UNSPECIFIED";
    LogRecordFlags[LogRecordFlags["LOG_RECORD_FLAG_TRACE_FLAGS_MASK"] = 255] = "LOG_RECORD_FLAG_TRACE_FLAGS_MASK";
    LogRecordFlags[LogRecordFlags["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
})(LogRecordFlags || (exports.LogRecordFlags = LogRecordFlags = {}));
function logRecordFlagsFromJSON(object) {
    switch (object) {
        case 0:
        case 'LOG_RECORD_FLAG_UNSPECIFIED':
            return LogRecordFlags.LOG_RECORD_FLAG_UNSPECIFIED;
        case 255:
        case 'LOG_RECORD_FLAG_TRACE_FLAGS_MASK':
            return LogRecordFlags.LOG_RECORD_FLAG_TRACE_FLAGS_MASK;
        case -1:
        case 'UNRECOGNIZED':
        default:
            return LogRecordFlags.UNRECOGNIZED;
    }
}
exports.logRecordFlagsFromJSON = logRecordFlagsFromJSON;
function logRecordFlagsToJSON(object) {
    switch (object) {
        case LogRecordFlags.LOG_RECORD_FLAG_UNSPECIFIED:
            return 'LOG_RECORD_FLAG_UNSPECIFIED';
        case LogRecordFlags.LOG_RECORD_FLAG_TRACE_FLAGS_MASK:
            return 'LOG_RECORD_FLAG_TRACE_FLAGS_MASK';
        case LogRecordFlags.UNRECOGNIZED:
        default:
            return 'UNRECOGNIZED';
    }
}
exports.logRecordFlagsToJSON = logRecordFlagsToJSON;
function createBaseLogsData() {
    return { resourceLogs: [] };
}
exports.LogsData = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.resourceLogs !== undefined &&
            message.resourceLogs.length !== 0) {
            for (const v of message.resourceLogs) {
                exports.ResourceLogs.encode(v, writer.uint32(10).fork()).ldelim();
            }
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseLogsData();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.resourceLogs.push(exports.ResourceLogs.decode(reader, reader.uint32()));
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            resourceLogs: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.resourceLogs)
                ? object.resourceLogs.map((e) => exports.ResourceLogs.fromJSON(e))
                : [],
        };
    },
    toJSON(message) {
        var _a;
        const obj = {};
        if ((_a = message.resourceLogs) === null || _a === void 0 ? void 0 : _a.length) {
            obj.resourceLogs = message.resourceLogs.map((e) => exports.ResourceLogs.toJSON(e));
        }
        return obj;
    },
    create(base) {
        return exports.LogsData.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a;
        const message = createBaseLogsData();
        message.resourceLogs =
            ((_a = object.resourceLogs) === null || _a === void 0 ? void 0 : _a.map((e) => exports.ResourceLogs.fromPartial(e))) || [];
        return message;
    },
};
function createBaseResourceLogs() {
    return {
        resource: undefined,
        scopeLogs: [],
        instrumentationLibraryLogs: [],
        schemaUrl: '',
    };
}
exports.ResourceLogs = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.resource !== undefined) {
            resource_1.Resource.encode(message.resource, writer.uint32(10).fork()).ldelim();
        }
        if (message.scopeLogs !== undefined && message.scopeLogs.length !== 0) {
            for (const v of message.scopeLogs) {
                exports.ScopeLogs.encode(v, writer.uint32(18).fork()).ldelim();
            }
        }
        if (message.instrumentationLibraryLogs !== undefined &&
            message.instrumentationLibraryLogs.length !== 0) {
            for (const v of message.instrumentationLibraryLogs) {
                exports.InstrumentationLibraryLogs.encode(v, writer.uint32(8002).fork()).ldelim();
            }
        }
        if (message.schemaUrl !== undefined && message.schemaUrl !== '') {
            writer.uint32(26).string(message.schemaUrl);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseResourceLogs();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.resource = resource_1.Resource.decode(reader, reader.uint32());
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.scopeLogs.push(exports.ScopeLogs.decode(reader, reader.uint32()));
                    continue;
                case 1000:
                    if (tag !== 8002) {
                        break;
                    }
                    message.instrumentationLibraryLogs.push(exports.InstrumentationLibraryLogs.decode(reader, reader.uint32()));
                    continue;
                case 3:
                    if (tag !== 26) {
                        break;
                    }
                    message.schemaUrl = reader.string();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            resource: isSet(object.resource)
                ? resource_1.Resource.fromJSON(object.resource)
                : undefined,
            scopeLogs: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.scopeLogs)
                ? object.scopeLogs.map((e) => exports.ScopeLogs.fromJSON(e))
                : [],
            instrumentationLibraryLogs: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.instrumentationLibraryLogs)
                ? object.instrumentationLibraryLogs.map((e) => exports.InstrumentationLibraryLogs.fromJSON(e))
                : [],
            schemaUrl: isSet(object.schemaUrl)
                ? globalThis.String(object.schemaUrl)
                : '',
        };
    },
    toJSON(message) {
        var _a, _b;
        const obj = {};
        if (message.resource !== undefined) {
            obj.resource = resource_1.Resource.toJSON(message.resource);
        }
        if ((_a = message.scopeLogs) === null || _a === void 0 ? void 0 : _a.length) {
            obj.scopeLogs = message.scopeLogs.map((e) => exports.ScopeLogs.toJSON(e));
        }
        if ((_b = message.instrumentationLibraryLogs) === null || _b === void 0 ? void 0 : _b.length) {
            obj.instrumentationLibraryLogs = message.instrumentationLibraryLogs.map((e) => exports.InstrumentationLibraryLogs.toJSON(e));
        }
        if (message.schemaUrl !== undefined && message.schemaUrl !== '') {
            obj.schemaUrl = message.schemaUrl;
        }
        return obj;
    },
    create(base) {
        return exports.ResourceLogs.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b, _c;
        const message = createBaseResourceLogs();
        message.resource =
            object.resource !== undefined && object.resource !== null
                ? resource_1.Resource.fromPartial(object.resource)
                : undefined;
        message.scopeLogs =
            ((_a = object.scopeLogs) === null || _a === void 0 ? void 0 : _a.map((e) => exports.ScopeLogs.fromPartial(e))) || [];
        message.instrumentationLibraryLogs =
            ((_b = object.instrumentationLibraryLogs) === null || _b === void 0 ? void 0 : _b.map((e) => exports.InstrumentationLibraryLogs.fromPartial(e))) || [];
        message.schemaUrl = (_c = object.schemaUrl) !== null && _c !== void 0 ? _c : '';
        return message;
    },
};
function createBaseScopeLogs() {
    return { scope: undefined, logRecords: [], schemaUrl: '' };
}
exports.ScopeLogs = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.scope !== undefined) {
            common_1.InstrumentationScope.encode(message.scope, writer.uint32(10).fork()).ldelim();
        }
        if (message.logRecords !== undefined && message.logRecords.length !== 0) {
            for (const v of message.logRecords) {
                exports.LogRecord.encode(v, writer.uint32(18).fork()).ldelim();
            }
        }
        if (message.schemaUrl !== undefined && message.schemaUrl !== '') {
            writer.uint32(26).string(message.schemaUrl);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseScopeLogs();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.scope = common_1.InstrumentationScope.decode(reader, reader.uint32());
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.logRecords.push(exports.LogRecord.decode(reader, reader.uint32()));
                    continue;
                case 3:
                    if (tag !== 26) {
                        break;
                    }
                    message.schemaUrl = reader.string();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            scope: isSet(object.scope)
                ? common_1.InstrumentationScope.fromJSON(object.scope)
                : undefined,
            logRecords: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.logRecords)
                ? object.logRecords.map((e) => exports.LogRecord.fromJSON(e))
                : [],
            schemaUrl: isSet(object.schemaUrl)
                ? globalThis.String(object.schemaUrl)
                : '',
        };
    },
    toJSON(message) {
        var _a;
        const obj = {};
        if (message.scope !== undefined) {
            obj.scope = common_1.InstrumentationScope.toJSON(message.scope);
        }
        if ((_a = message.logRecords) === null || _a === void 0 ? void 0 : _a.length) {
            obj.logRecords = message.logRecords.map((e) => exports.LogRecord.toJSON(e));
        }
        if (message.schemaUrl !== undefined && message.schemaUrl !== '') {
            obj.schemaUrl = message.schemaUrl;
        }
        return obj;
    },
    create(base) {
        return exports.ScopeLogs.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b;
        const message = createBaseScopeLogs();
        message.scope =
            object.scope !== undefined && object.scope !== null
                ? common_1.InstrumentationScope.fromPartial(object.scope)
                : undefined;
        message.logRecords =
            ((_a = object.logRecords) === null || _a === void 0 ? void 0 : _a.map((e) => exports.LogRecord.fromPartial(e))) || [];
        message.schemaUrl = (_b = object.schemaUrl) !== null && _b !== void 0 ? _b : '';
        return message;
    },
};
function createBaseInstrumentationLibraryLogs() {
    return { instrumentationLibrary: undefined, logRecords: [], schemaUrl: '' };
}
exports.InstrumentationLibraryLogs = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.instrumentationLibrary !== undefined) {
            common_1.InstrumentationLibrary.encode(message.instrumentationLibrary, writer.uint32(10).fork()).ldelim();
        }
        if (message.logRecords !== undefined && message.logRecords.length !== 0) {
            for (const v of message.logRecords) {
                exports.LogRecord.encode(v, writer.uint32(18).fork()).ldelim();
            }
        }
        if (message.schemaUrl !== undefined && message.schemaUrl !== '') {
            writer.uint32(26).string(message.schemaUrl);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseInstrumentationLibraryLogs();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.instrumentationLibrary = common_1.InstrumentationLibrary.decode(reader, reader.uint32());
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.logRecords.push(exports.LogRecord.decode(reader, reader.uint32()));
                    continue;
                case 3:
                    if (tag !== 26) {
                        break;
                    }
                    message.schemaUrl = reader.string();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            instrumentationLibrary: isSet(object.instrumentationLibrary)
                ? common_1.InstrumentationLibrary.fromJSON(object.instrumentationLibrary)
                : undefined,
            logRecords: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.logRecords)
                ? object.logRecords.map((e) => exports.LogRecord.fromJSON(e))
                : [],
            schemaUrl: isSet(object.schemaUrl)
                ? globalThis.String(object.schemaUrl)
                : '',
        };
    },
    toJSON(message) {
        var _a;
        const obj = {};
        if (message.instrumentationLibrary !== undefined) {
            obj.instrumentationLibrary = common_1.InstrumentationLibrary.toJSON(message.instrumentationLibrary);
        }
        if ((_a = message.logRecords) === null || _a === void 0 ? void 0 : _a.length) {
            obj.logRecords = message.logRecords.map((e) => exports.LogRecord.toJSON(e));
        }
        if (message.schemaUrl !== undefined && message.schemaUrl !== '') {
            obj.schemaUrl = message.schemaUrl;
        }
        return obj;
    },
    create(base) {
        return exports.InstrumentationLibraryLogs.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b;
        const message = createBaseInstrumentationLibraryLogs();
        message.instrumentationLibrary =
            object.instrumentationLibrary !== undefined &&
                object.instrumentationLibrary !== null
                ? common_1.InstrumentationLibrary.fromPartial(object.instrumentationLibrary)
                : undefined;
        message.logRecords =
            ((_a = object.logRecords) === null || _a === void 0 ? void 0 : _a.map((e) => exports.LogRecord.fromPartial(e))) || [];
        message.schemaUrl = (_b = object.schemaUrl) !== null && _b !== void 0 ? _b : '';
        return message;
    },
};
function createBaseLogRecord() {
    return {
        timeUnixNano: 0,
        observedTimeUnixNano: 0,
        severityNumber: 0,
        severityText: '',
        body: undefined,
        attributes: [],
        droppedAttributesCount: 0,
        flags: 0,
        traceId: new Uint8Array(0),
        spanId: new Uint8Array(0),
    };
}
exports.LogRecord = {
    encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.timeUnixNano !== undefined && message.timeUnixNano !== 0) {
            writer.uint32(9).fixed64(message.timeUnixNano);
        }
        if (message.observedTimeUnixNano !== undefined &&
            message.observedTimeUnixNano !== 0) {
            writer.uint32(89).fixed64(message.observedTimeUnixNano);
        }
        if (message.severityNumber !== undefined && message.severityNumber !== 0) {
            writer.uint32(16).int32(message.severityNumber);
        }
        if (message.severityText !== undefined && message.severityText !== '') {
            writer.uint32(26).string(message.severityText);
        }
        if (message.body !== undefined) {
            common_1.AnyValue.encode(message.body, writer.uint32(42).fork()).ldelim();
        }
        if (message.attributes !== undefined && message.attributes.length !== 0) {
            for (const v of message.attributes) {
                common_1.KeyValue.encode(v, writer.uint32(50).fork()).ldelim();
            }
        }
        if (message.droppedAttributesCount !== undefined &&
            message.droppedAttributesCount !== 0) {
            writer.uint32(56).uint32(message.droppedAttributesCount);
        }
        if (message.flags !== undefined && message.flags !== 0) {
            writer.uint32(69).fixed32(message.flags);
        }
        if (message.traceId !== undefined && message.traceId.length !== 0) {
            writer.uint32(74).bytes(message.traceId);
        }
        if (message.spanId !== undefined && message.spanId.length !== 0) {
            writer.uint32(82).bytes(message.spanId);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseLogRecord();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 9) {
                        break;
                    }
                    message.timeUnixNano = longToNumber(reader.fixed64());
                    continue;
                case 11:
                    if (tag !== 89) {
                        break;
                    }
                    message.observedTimeUnixNano = longToNumber(reader.fixed64());
                    continue;
                case 2:
                    if (tag !== 16) {
                        break;
                    }
                    message.severityNumber = reader.int32();
                    continue;
                case 3:
                    if (tag !== 26) {
                        break;
                    }
                    message.severityText = reader.string();
                    continue;
                case 5:
                    if (tag !== 42) {
                        break;
                    }
                    message.body = common_1.AnyValue.decode(reader, reader.uint32());
                    continue;
                case 6:
                    if (tag !== 50) {
                        break;
                    }
                    message.attributes.push(common_1.KeyValue.decode(reader, reader.uint32()));
                    continue;
                case 7:
                    if (tag !== 56) {
                        break;
                    }
                    message.droppedAttributesCount = reader.uint32();
                    continue;
                case 8:
                    if (tag !== 69) {
                        break;
                    }
                    message.flags = reader.fixed32();
                    continue;
                case 9:
                    if (tag !== 74) {
                        break;
                    }
                    message.traceId = reader.bytes();
                    continue;
                case 10:
                    if (tag !== 82) {
                        break;
                    }
                    message.spanId = reader.bytes();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            timeUnixNano: isSet(object.timeUnixNano)
                ? globalThis.Number(object.timeUnixNano)
                : 0,
            observedTimeUnixNano: isSet(object.observedTimeUnixNano)
                ? globalThis.Number(object.observedTimeUnixNano)
                : 0,
            severityNumber: isSet(object.severityNumber)
                ? severityNumberFromJSON(object.severityNumber)
                : 0,
            severityText: isSet(object.severityText)
                ? globalThis.String(object.severityText)
                : '',
            body: isSet(object.body) ? common_1.AnyValue.fromJSON(object.body) : undefined,
            attributes: globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.attributes)
                ? object.attributes.map((e) => common_1.KeyValue.fromJSON(e))
                : [],
            droppedAttributesCount: isSet(object.droppedAttributesCount)
                ? globalThis.Number(object.droppedAttributesCount)
                : 0,
            flags: isSet(object.flags) ? globalThis.Number(object.flags) : 0,
            traceId: isSet(object.traceId)
                ? bytesFromBase64(object.traceId)
                : new Uint8Array(0),
            spanId: isSet(object.spanId)
                ? bytesFromBase64(object.spanId)
                : new Uint8Array(0),
        };
    },
    toJSON(message) {
        var _a;
        const obj = {};
        if (message.timeUnixNano !== undefined && message.timeUnixNano !== 0) {
            obj.timeUnixNano = Math.round(message.timeUnixNano);
        }
        if (message.observedTimeUnixNano !== undefined &&
            message.observedTimeUnixNano !== 0) {
            obj.observedTimeUnixNano = Math.round(message.observedTimeUnixNano);
        }
        if (message.severityNumber !== undefined && message.severityNumber !== 0) {
            obj.severityNumber = severityNumberToJSON(message.severityNumber);
        }
        if (message.severityText !== undefined && message.severityText !== '') {
            obj.severityText = message.severityText;
        }
        if (message.body !== undefined) {
            obj.body = common_1.AnyValue.toJSON(message.body);
        }
        if ((_a = message.attributes) === null || _a === void 0 ? void 0 : _a.length) {
            obj.attributes = message.attributes.map((e) => common_1.KeyValue.toJSON(e));
        }
        if (message.droppedAttributesCount !== undefined &&
            message.droppedAttributesCount !== 0) {
            obj.droppedAttributesCount = Math.round(message.droppedAttributesCount);
        }
        if (message.flags !== undefined && message.flags !== 0) {
            obj.flags = Math.round(message.flags);
        }
        if (message.traceId !== undefined && message.traceId.length !== 0) {
            obj.traceId = base64FromBytes(message.traceId);
        }
        if (message.spanId !== undefined && message.spanId.length !== 0) {
            obj.spanId = base64FromBytes(message.spanId);
        }
        return obj;
    },
    create(base) {
        return exports.LogRecord.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j;
        const message = createBaseLogRecord();
        message.timeUnixNano = (_a = object.timeUnixNano) !== null && _a !== void 0 ? _a : 0;
        message.observedTimeUnixNano = (_b = object.observedTimeUnixNano) !== null && _b !== void 0 ? _b : 0;
        message.severityNumber = (_c = object.severityNumber) !== null && _c !== void 0 ? _c : 0;
        message.severityText = (_d = object.severityText) !== null && _d !== void 0 ? _d : '';
        message.body =
            object.body !== undefined && object.body !== null
                ? common_1.AnyValue.fromPartial(object.body)
                : undefined;
        message.attributes =
            ((_e = object.attributes) === null || _e === void 0 ? void 0 : _e.map((e) => common_1.KeyValue.fromPartial(e))) || [];
        message.droppedAttributesCount = (_f = object.droppedAttributesCount) !== null && _f !== void 0 ? _f : 0;
        message.flags = (_g = object.flags) !== null && _g !== void 0 ? _g : 0;
        message.traceId = (_h = object.traceId) !== null && _h !== void 0 ? _h : new Uint8Array(0);
        message.spanId = (_j = object.spanId) !== null && _j !== void 0 ? _j : new Uint8Array(0);
        return message;
    },
};
function bytesFromBase64(b64) {
    if (globalThis.Buffer) {
        return Uint8Array.from(globalThis.Buffer.from(b64, 'base64'));
    }
    else {
        const bin = globalThis.atob(b64);
        const arr = new Uint8Array(bin.length);
        for (let i = 0; i < bin.length; ++i) {
            arr[i] = bin.charCodeAt(i);
        }
        return arr;
    }
}
function base64FromBytes(arr) {
    if (globalThis.Buffer) {
        return globalThis.Buffer.from(arr).toString('base64');
    }
    else {
        const bin = [];
        arr.forEach((byte) => {
            bin.push(globalThis.String.fromCharCode(byte));
        });
        return globalThis.btoa(bin.join(''));
    }
}
function longToNumber(long) {
    if (long.gt(globalThis.Number.MAX_SAFE_INTEGER)) {
        throw new globalThis.Error('Value is larger than Number.MAX_SAFE_INTEGER');
    }
    return long.toNumber();
}
if (minimal_1.default.util.Long !== long_1.default) {
    minimal_1.default.util.Long = long_1.default;
    minimal_1.default.configure();
}
function isSet(value) {
    return value !== null && value !== undefined;
}
