// @hyperdx/otel-web-session-recorder/dist/cjs/index.js
"use strict";
/*
Copyright 2023 DeploySentinel, Inc.
Copyright 2022 Splunk Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
const api_1 = require("@opentelemetry/api");
const rrweb_1 = require("rrweb");
const OTLPLogExporter_1 = require("./OTLPLogExporter");
const BatchLogProcessor_1 = require("./BatchLogProcessor");
const version_1 = require("./version");
const sessionrecording_utils_1 = require("./sessionrecording-utils");
// Hard limit of 4 hours of maximum recording during one session
const MAX_RECORDING_LENGTH = (4 * 60 + 1) * 60 * 1000;
const MAX_CHUNK_SIZE = 950 * 1024; // ~950KB
const encoder = new TextEncoder();
const decoder = new TextDecoder();
let inited = false;
let tracer;
let lastKnownSession;
let sessionStartTime = 0;
let paused = false;
let eventCounter = 1;
let logCounter = 1;
let mutationRateLimiter;
let queuedFullSnapshotTimeout;
const RumRecorder = {
    get inited() {
        return Boolean(inited);
    },
    init(config) {
        if (inited) {
            return;
        }
        if (typeof window === 'undefined') {
            console.error(
            // eslint-disable-next-line quotes
            "Session recorder can't be ran in non-browser environments");
            return;
        }
        let tracerProvider = api_1.trace.getTracerProvider();
        if (tracerProvider && 'getDelegate' in tracerProvider) {
            tracerProvider = tracerProvider.getDelegate();
        }
        if (!(tracerProvider === null || tracerProvider === void 0 ? void 0 : tracerProvider.resource)) {
            console.error('RUM OTEL Web must be inited before recorder.');
            return;
        }
        const resource = tracerProvider.resource;
        const { apiKey, url, debug } = config, rrwebConf = __rest(config, ["apiKey", "url", "debug"]);
        tracer = api_1.trace.getTracer('rum.rr-web', version_1.VERSION);
        const span = tracer.startSpan('record init');
        // Check if sampler is ignoring this
        if (!span.isRecording()) {
            return;
        }
        span.end();
        const exportUrl = url !== null && url !== void 0 ? url : 'https://in-otel.hyperdx.io/v1/logs';
        const headers = {};
        if (apiKey) {
            headers['authorization'] = `${apiKey}`;
        }
        const exporter = new OTLPLogExporter_1.default({
            beaconUrl: exportUrl,
            debug,
            headers,
            resource,
        });
        const processor = new BatchLogProcessor_1.BatchLogProcessor(exporter, {});
        lastKnownSession = resource.attributes['rum.sessionId'];
        sessionStartTime = Date.now();
        mutationRateLimiter =
            mutationRateLimiter !== null && mutationRateLimiter !== void 0 ? mutationRateLimiter : new sessionrecording_utils_1.MutationRateLimiter(rrweb_1.record, {
                onBlockedNode: (id, node) => {
                    const message = `OpenTelemetry Session Recorder: Rate limiting recording mutations on node ${id}. This can be due to excessive animations.`;
                    // Construct a span for the warning message
                    const now = Date.now();
                    const span = tracer.startSpan('console.warn', {
                        kind: api_1.SpanKind.INTERNAL,
                        startTime: now,
                    });
                    span.setAttribute('component', 'console');
                    span.setAttribute('level', 'warn');
                    span.setAttribute('message', message);
                    span.end(now);
                    if (debug) {
                        console.warn(message, node);
                    }
                    clearTimeout(queuedFullSnapshotTimeout);
                    queuedFullSnapshotTimeout = setTimeout(() => {
                        // Take a full snapshot to capture more accurate final mutation state
                        // This is only fired once per node, so we can still miss the outcome
                        // of repeated mutations
                        rrweb_1.record.takeFullSnapshot();
                    }, 1000);
                },
            });
        inited = (0, rrweb_1.record)(Object.assign(Object.assign({ maskAllInputs: true, maskTextSelector: '*' }, rrwebConf), { emit(srcEvent) {
                if (paused) {
                    return;
                }
                // Safeguards from our ingest getting DDOSed:
                // 1. A session can send up to 4 hours of data
                // 2. Recording resumes on session change if it isn't a background tab (session regenerated in an another tab)
                if (resource.attributes['rum.sessionId'] !== lastKnownSession) {
                    if (document.hidden) {
                        return;
                    }
                    lastKnownSession = resource.attributes['rum.sessionId'];
                    sessionStartTime = Date.now();
                    // reset counters
                    eventCounter = 1;
                    logCounter = 1;
                    rrweb_1.record.takeFullSnapshot();
                }
                if (srcEvent.timestamp > sessionStartTime + MAX_RECORDING_LENGTH) {
                    return;
                }
                const event = mutationRateLimiter
                    ? mutationRateLimiter.throttleMutations(srcEvent)
                    : srcEvent;
                if (!event) {
                    return;
                }
                const time = event.timestamp;
                const eventI = eventCounter++;
                // Research found that stringifying the rr-web event here is
                // more efficient for otlp + gzip exporting
                // Blob is unicode aware for size calculation (eg emoji.length = 1 vs blob.size() = 4)
                const body = encoder.encode((0, sessionrecording_utils_1.ensureStringifiedMaxMessageSize)(JSON.stringify(event)));
                const totalC = Math.ceil(body.byteLength / MAX_CHUNK_SIZE);
                for (let i = 0; i < totalC; i++) {
                    const start = i * MAX_CHUNK_SIZE;
                    const end = (i + 1) * MAX_CHUNK_SIZE;
                    const log = (0, BatchLogProcessor_1.convert)(decoder.decode(body.slice(start, end)), time, {
                        'rr-web.offset': logCounter++,
                        'rr-web.event': eventI,
                        'rr-web.chunk': i + 1,
                        'rr-web.total-chunks': totalC,
                    });
                    if (debug) {
                        console.log(log);
                    }
                    processor.onLog(log);
                }
            } }));
    },
    resume() {
        if (!inited) {
            return;
        }
        const oldPaused = paused;
        paused = false;
        if (!oldPaused) {
            rrweb_1.record.takeFullSnapshot();
            tracer.startSpan('record resume').end();
        }
    },
    stop() {
        if (!inited) {
            return;
        }
        if (paused) {
            tracer.startSpan('record stop').end();
        }
        paused = true;
    },
    deinit() {
        if (!inited) {
            return;
        }
        inited();
        inited = false;
    },
};
exports.default = RumRecorder;
