// @hyperdx/otel-web-session-recorder/dist/cjs/OTLPLogExporter.js
"use strict";
/*
Copyright 2022 Splunk Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
Object.defineProperty(exports, "__esModule", { value: true });
const fflate_1 = require("fflate");
const proto = require("./opentelemetry/proto/logs/v1/logs");
const version_1 = require("./version");
const defaultHeaders = {
    'Content-Type': 'application/x-protobuf',
    'Content-Encoding': 'gzip',
};
const LogsData = proto.LogsData;
function isArray(value) {
    return Array.isArray(value);
}
function isObject(value) {
    return !!value && typeof value === 'object' && !isArray(value);
}
function convertToAnyValue(value) {
    if (isObject(value)) {
        return {
            kvlistValue: {
                values: Object.entries(value).map(([key, value]) => ({
                    key,
                    value: convertToAnyValue(value),
                })),
            },
        };
    }
    if (isArray(value)) {
        return {
            arrayValue: {
                values: value.map((value) => convertToAnyValue(value)),
            },
        };
    }
    if (typeof value === 'string') {
        return {
            stringValue: value,
        };
    }
    if (typeof value === 'number') {
        return {
            doubleValue: value,
        };
    }
    if (typeof value === 'boolean') {
        return {
            boolValue: value,
        };
    }
    // never
    return {};
}
// TODO: should use opentelemetry logs exporter instead (https://github.com/open-telemetry/opentelemetry-js/tree/main/experimental/packages/exporter-logs-otlp-proto)
class OTLPLogExporter {
    constructor(config) {
        this.config = config;
    }
    constructLogData(logs) {
        var _a, _b, _c;
        return {
            resourceLogs: [
                {
                    resource: {
                        attributes: (_c = (_b = convertToAnyValue(((_a = this.config.resource) === null || _a === void 0 ? void 0 : _a.attributes) || {})) === null || _b === void 0 ? void 0 : _b.kvlistValue) === null || _c === void 0 ? void 0 : _c.values,
                    },
                    scopeLogs: [
                        {
                            scope: { name: 'rum.rr-web', version: version_1.VERSION },
                            logRecords: logs.map((log) => {
                                var _a;
                                return ({
                                    body: convertToAnyValue(log.body),
                                    timeUnixNano: log.timeUnixNano * 1000000,
                                    attributes: (_a = convertToAnyValue(log.attributes || {}).kvlistValue) === null || _a === void 0 ? void 0 : _a.values,
                                });
                            }),
                        },
                    ],
                },
            ],
        };
    }
    export(logs) {
        if (logs.length === 0) {
            return;
        }
        const logsData = this.constructLogData(logs);
        if (this.config.debug) {
            console.log('otlp request', logsData);
        }
        const buffer = LogsData.encode(logsData).finish();
        const compressed = (0, fflate_1.gzipSync)(buffer);
        const updatedHeaders = this.config.headers
            ? Object.assign({}, defaultHeaders, this.config.headers)
            : defaultHeaders;
        fetch(this.config.beaconUrl, {
            method: 'POST',
            body: compressed,
            headers: updatedHeaders,
        }).catch(() => {
            // TODO remove this once we have ingest with correct cors headers
        });
    }
}
exports.default = OTLPLogExporter;
