// @hyperdx/otel-web-session-recorder/dist/cjs/sessionrecording-utils.js
"use strict";
/**
Copyright 2023 DeploySentinel, Inc.

Copyright 2020 Posthog / Hiberly, Inc.

Copyright 2015 Mixpanel, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this work except in compliance with the License.
You may obtain a copy of the License below, or at:

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

Source: https://github.com/PostHog/posthog-js/blob/b6e3855a7f4339fba011e9a747be67c3a9c78955/src/extensions/sessionrecording-utils.ts
*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.MutationRateLimiter = exports.ensureStringifiedMaxMessageSize = exports.MAX_MESSAGE_SIZE = exports.MUTATION_SOURCE_TYPE = exports.PLUGIN_EVENT_TYPE = exports.INCREMENTAL_SNAPSHOT_EVENT_TYPE = exports.META_EVENT_TYPE = exports.FULL_SNAPSHOT_EVENT_TYPE = exports.replacementImageURI = void 0;
exports.replacementImageURI = 'data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMTYiIGhlaWdodD0iMTYiIHZpZXdCb3g9IjAgMCAxNiAxNiIgZmlsbD0ibm9uZSIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIj4KPHJlY3Qgd2lkdGg9IjE2IiBoZWlnaHQ9IjE2IiBmaWxsPSJibGFjayIvPgo8cGF0aCBkPSJNOCAwSDE2TDAgMTZWOEw4IDBaIiBmaWxsPSIjMkQyRDJEIi8+CjxwYXRoIGQ9Ik0xNiA4VjE2SDhMMTYgOFoiIGZpbGw9IiMyRDJEMkQiLz4KPC9zdmc+Cg==';
exports.FULL_SNAPSHOT_EVENT_TYPE = 2;
exports.META_EVENT_TYPE = 4;
exports.INCREMENTAL_SNAPSHOT_EVENT_TYPE = 3;
exports.PLUGIN_EVENT_TYPE = 6;
exports.MUTATION_SOURCE_TYPE = 0;
exports.MAX_MESSAGE_SIZE = 5000000; // ~5mb
/*
 * Check whether a data payload is nearing 5mb. If it is, it checks the data for
 * data URIs (the likely culprit for large payloads). If it finds data URIs, it either replaces
 * it with a generic image (if it's an image) or removes it.
 * @data {object} the rr-web data object
 * @returns {object} the rr-web data object with data uris filtered out
 */
function ensureStringifiedMaxMessageSize(stringifiedData) {
    // Note: with compression, this limit may be able to be increased
    // but we're assuming most of the size is from a data uri which
    // is unlikely to be compressed further
    if (stringifiedData.length > exports.MAX_MESSAGE_SIZE) {
        // Regex that matches the pattern for a dataURI with the shape 'data:{mime type};{encoding},{data}'. It:
        // 1) Checks if the pattern starts with 'data:' (potentially, not at the start of the string)
        // 2) Extracts the mime type of the data uri in the first group
        // 3) Determines when the data URI ends.Depending on if it's used in the src tag or css, it can end with a ) or "
        const dataURIRegex = /data:([\w/\-.]+);(\w+),([^)"]*)/gim;
        const matches = stringifiedData.matchAll(dataURIRegex);
        for (const match of matches) {
            if (match[1].toLocaleLowerCase().slice(0, 6) === 'image/') {
                stringifiedData = stringifiedData.replace(match[0], exports.replacementImageURI);
            }
            else {
                stringifiedData = stringifiedData.replace(match[0], '');
            }
        }
    }
    return stringifiedData;
}
exports.ensureStringifiedMaxMessageSize = ensureStringifiedMaxMessageSize;
class MutationRateLimiter {
    constructor(rrweb, options = {}) {
        var _a, _b;
        this.rrweb = rrweb;
        this.options = options;
        this.bucketSize = 100;
        this.refillRate = 10;
        this.mutationBuckets = {};
        this.loggedTracker = {};
        this.refillBuckets = () => {
            Object.keys(this.mutationBuckets).forEach((key) => {
                this.mutationBuckets[key] = this.mutationBuckets[key] + this.refillRate;
                if (this.mutationBuckets[key] >= this.bucketSize) {
                    delete this.mutationBuckets[key];
                }
            });
        };
        this.getNodeOrRelevantParent = (id) => {
            // For some nodes we know they are part of a larger tree such as an SVG.
            // For those we want to block the entire node, not just the specific attribute
            const node = this.rrweb.mirror.getNode(id);
            // Check if the node is an Element and then find the closest parent that is an SVG
            if ((node === null || node === void 0 ? void 0 : node.nodeName) !== 'svg' && node instanceof Element) {
                const closestSVG = node.closest('svg');
                if (closestSVG) {
                    // @ts-ignore
                    return [this.rrweb.mirror.getId(closestSVG), closestSVG];
                }
            }
            return [id, node];
        };
        this.numberOfChanges = (data) => {
            var _a, _b, _c, _d, _e, _f, _g, _h;
            return (((_b = (_a = data.removes) === null || _a === void 0 ? void 0 : _a.length) !== null && _b !== void 0 ? _b : 0) +
                ((_d = (_c = data.attributes) === null || _c === void 0 ? void 0 : _c.length) !== null && _d !== void 0 ? _d : 0) +
                ((_f = (_e = data.texts) === null || _e === void 0 ? void 0 : _e.length) !== null && _f !== void 0 ? _f : 0) +
                ((_h = (_g = data.adds) === null || _g === void 0 ? void 0 : _g.length) !== null && _h !== void 0 ? _h : 0));
        };
        this.throttleMutations = (event) => {
            if (event.type !== exports.INCREMENTAL_SNAPSHOT_EVENT_TYPE ||
                event.data.source !== exports.MUTATION_SOURCE_TYPE) {
                return event;
            }
            const data = event.data;
            const initialMutationCount = this.numberOfChanges(data);
            if (data.attributes) {
                // Most problematic mutations come from attrs where the style or minor properties are changed rapidly
                data.attributes = data.attributes.filter((attr) => {
                    var _a, _b, _c;
                    const [nodeId, node] = this.getNodeOrRelevantParent(attr.id);
                    if (this.mutationBuckets[nodeId] === 0) {
                        return false;
                    }
                    this.mutationBuckets[nodeId] =
                        (_a = this.mutationBuckets[nodeId]) !== null && _a !== void 0 ? _a : this.bucketSize;
                    this.mutationBuckets[nodeId] = Math.max(this.mutationBuckets[nodeId] - 1, 0);
                    if (this.mutationBuckets[nodeId] === 0) {
                        if (!this.loggedTracker[nodeId]) {
                            this.loggedTracker[nodeId] = true;
                            (_c = (_b = this.options).onBlockedNode) === null || _c === void 0 ? void 0 : _c.call(_b, nodeId, node);
                        }
                    }
                    return attr;
                });
            }
            // Check if every part of the mutation is empty in which case there is nothing to do
            const mutationCount = this.numberOfChanges(data);
            if (mutationCount === 0 && initialMutationCount !== mutationCount) {
                // If we have modified the mutation count and the remaining count is 0, then we don't need the event.
                return;
            }
            return event;
        };
        this.refillRate = (_a = this.options.refillRate) !== null && _a !== void 0 ? _a : this.refillRate;
        this.bucketSize = (_b = this.options.bucketSize) !== null && _b !== void 0 ? _b : this.bucketSize;
        setInterval(() => {
            this.refillBuckets();
        }, 1000);
    }
}
exports.MutationRateLimiter = MutationRateLimiter;
