// @hyperdx/instrumentation-exception/build/src/browser/eventbuilder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.eventFromUnknownInput = exports.eventFromMessage = exports.eventFromException = exports.exceptionFromError = void 0;
const core_1 = require("@sentry/core");
const utils_1 = require("@sentry/utils");
function exceptionFromError(stackParser, ex) {
    const frames = parseStackFrames(stackParser, ex);
    const exception = {
        type: ex && ex.name,
        value: extractMessage(ex),
    };
    if (frames.length) {
        exception.stacktrace = { frames };
    }
    if (exception.type === undefined && exception.value === '') {
        exception.value = 'Unrecoverable error caught';
    }
    return exception;
}
exports.exceptionFromError = exceptionFromError;
function eventFromPlainObject(stackParser, exception, syntheticException, isUnhandledRejection) {
    const client = (0, core_1.getClient)();
    const normalizeDepth = client && client.getOptions().normalizeDepth;
    const errorFromProp = getErrorPropertyFromObject(exception);
    const extra = {
        __serialized__: (0, utils_1.normalizeToSize)(exception, normalizeDepth),
    };
    if (errorFromProp) {
        return {
            exception: {
                values: [exceptionFromError(stackParser, errorFromProp)],
            },
            extra,
        };
    }
    const event = {
        exception: {
            values: [
                {
                    type: (0, utils_1.isEvent)(exception)
                        ? exception.constructor.name
                        : isUnhandledRejection
                            ? 'UnhandledRejection'
                            : 'Error',
                    value: getNonErrorObjectExceptionValue(exception, {
                        isUnhandledRejection,
                    }),
                },
            ],
        },
        extra,
    };
    if (syntheticException) {
        const frames = parseStackFrames(stackParser, syntheticException);
        if (frames.length) {
            event.exception.values[0].stacktrace = { frames };
        }
    }
    return event;
}
function eventFromError(stackParser, ex) {
    return {
        exception: {
            values: [exceptionFromError(stackParser, ex)],
        },
    };
}
function parseStackFrames(stackParser, ex) {
    const stacktrace = ex.stacktrace || ex.stack || '';
    const skipLines = getSkipFirstStackStringLines(ex);
    const framesToPop = getPopFirstTopFrames(ex);
    try {
        return stackParser(stacktrace, skipLines, framesToPop);
    }
    catch (e) {
    }
    return [];
}
const reactMinifiedRegexp = /Minified React error #\d+;/i;
function getSkipFirstStackStringLines(ex) {
    if (ex && reactMinifiedRegexp.test(ex.message)) {
        return 1;
    }
    return 0;
}
function getPopFirstTopFrames(ex) {
    if (typeof ex.framesToPop === 'number') {
        return ex.framesToPop;
    }
    return 0;
}
function extractMessage(ex) {
    const message = ex && ex.message;
    if (!message) {
        return 'No error message';
    }
    if (message.error && typeof message.error.message === 'string') {
        return message.error.message;
    }
    return message;
}
function eventFromException(stackParser, exception, hint, attachStacktrace) {
    const syntheticException = (hint && hint.syntheticException) || undefined;
    const event = eventFromUnknownInput(stackParser, exception, syntheticException, attachStacktrace);
    (0, utils_1.addExceptionMechanism)(event);
    event.level = 'error';
    if (hint && hint.event_id) {
        event.event_id = hint.event_id;
    }
    return (0, utils_1.resolvedSyncPromise)(event);
}
exports.eventFromException = eventFromException;
function eventFromMessage(stackParser, message, level = 'info', hint, attachStacktrace) {
    const syntheticException = (hint && hint.syntheticException) || undefined;
    const event = eventFromString(stackParser, message, syntheticException, attachStacktrace);
    event.level = level;
    if (hint && hint.event_id) {
        event.event_id = hint.event_id;
    }
    return (0, utils_1.resolvedSyncPromise)(event);
}
exports.eventFromMessage = eventFromMessage;
function eventFromUnknownInput(stackParser, exception, syntheticException, attachStacktrace, isUnhandledRejection) {
    let event;
    if ((0, utils_1.isErrorEvent)(exception) &&
        exception.error) {
        const errorEvent = exception;
        return eventFromError(stackParser, errorEvent.error);
    }
    if ((0, utils_1.isDOMError)(exception) || (0, utils_1.isDOMException)(exception)) {
        const domException = exception;
        if ('stack' in exception) {
            event = eventFromError(stackParser, exception);
        }
        else {
            const name = domException.name ||
                ((0, utils_1.isDOMError)(domException) ? 'DOMError' : 'DOMException');
            const message = domException.message
                ? `${name}: ${domException.message}`
                : name;
            event = eventFromString(stackParser, message, syntheticException, attachStacktrace);
            (0, utils_1.addExceptionTypeValue)(event, message);
        }
        if ('code' in domException) {
            event.tags = {
                ...event.tags,
                'DOMException.code': `${domException.code}`,
            };
        }
        return event;
    }
    if ((0, utils_1.isError)(exception)) {
        return eventFromError(stackParser, exception);
    }
    if ((0, utils_1.isPlainObject)(exception) || (0, utils_1.isEvent)(exception)) {
        const objectException = exception;
        event = eventFromPlainObject(stackParser, objectException, syntheticException, isUnhandledRejection);
        (0, utils_1.addExceptionMechanism)(event, {
            synthetic: true,
        });
        return event;
    }
    event = eventFromString(stackParser, exception, syntheticException, attachStacktrace);
    (0, utils_1.addExceptionTypeValue)(event, `${exception}`, undefined);
    (0, utils_1.addExceptionMechanism)(event, {
        synthetic: true,
    });
    return event;
}
exports.eventFromUnknownInput = eventFromUnknownInput;
function eventFromString(stackParser, message, syntheticException, attachStacktrace) {
    const event = {};
    if (attachStacktrace && syntheticException) {
        const frames = parseStackFrames(stackParser, syntheticException);
        if (frames.length) {
            event.exception = {
                values: [{ value: message, stacktrace: { frames } }],
            };
        }
    }
    if ((0, utils_1.isParameterizedString)(message)) {
        const { __sentry_template_string__, __sentry_template_values__ } = message;
        event.logentry = {
            message: __sentry_template_string__,
            params: __sentry_template_values__,
        };
        return event;
    }
    event.message = message;
    return event;
}
function getNonErrorObjectExceptionValue(exception, { isUnhandledRejection }) {
    const keys = (0, utils_1.extractExceptionKeysForMessage)(exception);
    const captureType = isUnhandledRejection ? 'promise rejection' : 'exception';
    if ((0, utils_1.isErrorEvent)(exception)) {
        return `Event \`ErrorEvent\` captured as ${captureType} with message \`${exception.message}\``;
    }
    if ((0, utils_1.isEvent)(exception)) {
        const className = getObjectClassName(exception);
        return `Event \`${className}\` (type=${exception.type}) captured as ${captureType}`;
    }
    return `Object captured as ${captureType} with keys: ${keys}`;
}
function getObjectClassName(obj) {
    try {
        const prototype = Object.getPrototypeOf(obj);
        return prototype ? prototype.constructor.name : undefined;
    }
    catch (e) {
    }
}
function getErrorPropertyFromObject(obj) {
    for (const prop in obj) {
        if (Object.prototype.hasOwnProperty.call(obj, prop)) {
            const value = obj[prop];
            if (value instanceof Error) {
                return value;
            }
        }
    }
    return undefined;
}
