// @hyperdx/instrumentation-exception/build/src/browser/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.recordException = exports.buildEventFromException = exports.applyIntegrations = exports.setupIntegrations = void 0;
const core_1 = require("@sentry/core");
const api_1 = require("@opentelemetry/api");
const instrumentation_sentry_node_1 = require("@hyperdx/instrumentation-sentry-node");
const browserapierrors_1 = require("./integrations/browserapierrors");
const contextlines_1 = require("./integrations/contextlines");
const stack_parsers_1 = require("./stack-parsers");
const eventbuilder_1 = require("./eventbuilder");
const httpcontext_1 = require("./integrations/httpcontext");
const hyperdx_1 = require("./integrations/hyperdx");
const linkederrors_1 = require("./integrations/linkederrors");
const package_json_1 = require("../../package.json");
const defaultTracer = api_1.trace.getTracer(package_json_1.name, package_json_1.version);
const defaultIntegrations = [
    (0, core_1.inboundFiltersIntegration)(),
    (0, contextlines_1.contextLinesIntegration)(),
    (0, browserapierrors_1.browserApiErrorsIntegration)(),
    (0, linkederrors_1.linkedErrorsIntegration)(),
    (0, httpcontext_1.httpContextIntegration)(),
    (0, hyperdx_1.hyperdxIntegration)(),
];
const DEFAULT_CLIENT_OPTIONS = {
    integrations: defaultIntegrations,
    stackParser: stack_parsers_1.defaultStackParser,
    maxValueLength: 250,
    transport: {},
};
const fakeClient = {
    getOptions: () => DEFAULT_CLIENT_OPTIONS,
};
const setupIntegrations = () => {
    var _a, _b;
    for (const integration of defaultIntegrations) {
        (_a = integration.setupOnce) === null || _a === void 0 ? void 0 : _a.call(integration);
        (_b = integration.setup) === null || _b === void 0 ? void 0 : _b.call(integration, fakeClient);
    }
};
exports.setupIntegrations = setupIntegrations;
const applyIntegrations = async (event, hint) => {
    var _a;
    for (const integration of defaultIntegrations) {
        await ((_a = integration.preprocessEvent) === null || _a === void 0 ? void 0 : _a.call(integration, event, hint, fakeClient));
    }
    for (const integration of defaultIntegrations) {
        if (typeof integration.processEvent === 'function') {
            event = await integration.processEvent(event, hint, fakeClient);
        }
    }
    return event;
};
exports.applyIntegrations = applyIntegrations;
const buildEventFromException = async (exception, hint) => {
    let event = (0, eventbuilder_1.eventFromUnknownInput)(stack_parsers_1.defaultStackParser, exception);
    try {
        event = await (0, core_1.prepareEvent)(DEFAULT_CLIENT_OPTIONS, event, hint);
    }
    catch (err) {
        api_1.diag.error('Failed to prepare event', err);
    }
    return (0, exports.applyIntegrations)(event, hint);
};
exports.buildEventFromException = buildEventFromException;
const recordException = async (e, hint) => {
    try {
        const { tracer, span, attributes, ...eventHint } = hint !== null && hint !== void 0 ? hint : {};
        const _hint = Object.keys(eventHint).length > 0
            ? eventHint
            : {
                mechanism: {
                    type: 'generic',
                    handled: true,
                },
            };
        const _eventProcessor = (0, instrumentation_sentry_node_1.getEventProcessor)(tracer !== null && tracer !== void 0 ? tracer : defaultTracer);
        const event = await (0, exports.buildEventFromException)(e, {
            data: _hint,
        });
        _eventProcessor(event, _hint, span, attributes);
    }
    catch (err) {
        api_1.diag.error('Failed to capture exception', err);
    }
};
exports.recordException = recordException;
