// @hyperdx/instrumentation-exception/build/src/browser/stack-parsers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.defaultStackParser = exports.defaultStackLineParsers = exports.opera11StackLineParser = exports.opera10StackLineParser = exports.winjsStackLineParser = exports.geckoStackLineParser = exports.chromeStackLineParser = void 0;
const utils_1 = require("@sentry/utils");
const OPERA10_PRIORITY = 10;
const OPERA11_PRIORITY = 20;
const CHROME_PRIORITY = 30;
const WINJS_PRIORITY = 40;
const GECKO_PRIORITY = 50;
function createFrame(filename, func, lineno, colno) {
    const frame = {
        filename,
        function: func === '<anonymous>' ? utils_1.UNKNOWN_FUNCTION : func,
        in_app: true,
    };
    if (lineno !== undefined) {
        frame.lineno = lineno;
    }
    if (colno !== undefined) {
        frame.colno = colno;
    }
    return frame;
}
const chromeRegex = /^\s*at (?:(.+?\)(?: \[.+\])?|.*?) ?\((?:address at )?)?(?:async )?((?:<anonymous>|[-a-z]+:|.*bundle|\/)?.*?)(?::(\d+))?(?::(\d+))?\)?\s*$/i;
const chromeEvalRegex = /\((\S*)(?::(\d+))(?::(\d+))\)/;
const chromeStackParserFn = (line) => {
    const parts = chromeRegex.exec(line);
    if (parts) {
        const isEval = parts[2] && parts[2].indexOf('eval') === 0;
        if (isEval) {
            const subMatch = chromeEvalRegex.exec(parts[2]);
            if (subMatch) {
                parts[2] = subMatch[1];
                parts[3] = subMatch[2];
                parts[4] = subMatch[3];
            }
        }
        const [func, filename] = extractSafariExtensionDetails(parts[1] || utils_1.UNKNOWN_FUNCTION, parts[2]);
        return createFrame(filename, func, parts[3] ? +parts[3] : undefined, parts[4] ? +parts[4] : undefined);
    }
    return;
};
exports.chromeStackLineParser = [
    CHROME_PRIORITY,
    chromeStackParserFn,
];
const geckoREgex = /^\s*(.*?)(?:\((.*?)\))?(?:^|@)?((?:[-a-z]+)?:\/.*?|\[native code\]|[^@]*(?:bundle|\d+\.js)|\/[\w\-. /=]+)(?::(\d+))?(?::(\d+))?\s*$/i;
const geckoEvalRegex = /(\S+) line (\d+)(?: > eval line \d+)* > eval/i;
const gecko = (line) => {
    const parts = geckoREgex.exec(line);
    if (parts) {
        const isEval = parts[3] && parts[3].indexOf(' > eval') > -1;
        if (isEval) {
            const subMatch = geckoEvalRegex.exec(parts[3]);
            if (subMatch) {
                parts[1] = parts[1] || 'eval';
                parts[3] = subMatch[1];
                parts[4] = subMatch[2];
                parts[5] = '';
            }
        }
        let filename = parts[3];
        let func = parts[1] || utils_1.UNKNOWN_FUNCTION;
        [func, filename] = extractSafariExtensionDetails(func, filename);
        return createFrame(filename, func, parts[4] ? +parts[4] : undefined, parts[5] ? +parts[5] : undefined);
    }
    return;
};
exports.geckoStackLineParser = [GECKO_PRIORITY, gecko];
const winjsRegex = /^\s*at (?:((?:\[object object\])?.+) )?\(?((?:[-a-z]+):.*?):(\d+)(?::(\d+))?\)?\s*$/i;
const winjs = (line) => {
    const parts = winjsRegex.exec(line);
    return parts
        ? createFrame(parts[2], parts[1] || utils_1.UNKNOWN_FUNCTION, +parts[3], parts[4] ? +parts[4] : undefined)
        : undefined;
};
exports.winjsStackLineParser = [WINJS_PRIORITY, winjs];
const opera10Regex = / line (\d+).*script (?:in )?(\S+)(?:: in function (\S+))?$/i;
const opera10 = (line) => {
    const parts = opera10Regex.exec(line);
    return parts
        ? createFrame(parts[2], parts[3] || utils_1.UNKNOWN_FUNCTION, +parts[1])
        : undefined;
};
exports.opera10StackLineParser = [
    OPERA10_PRIORITY,
    opera10,
];
const opera11Regex = / line (\d+), column (\d+)\s*(?:in (?:<anonymous function: ([^>]+)>|([^)]+))\(.*\))? in (.*):\s*$/i;
const opera11 = (line) => {
    const parts = opera11Regex.exec(line);
    return parts
        ? createFrame(parts[5], parts[3] || parts[4] || utils_1.UNKNOWN_FUNCTION, +parts[1], +parts[2])
        : undefined;
};
exports.opera11StackLineParser = [
    OPERA11_PRIORITY,
    opera11,
];
exports.defaultStackLineParsers = [
    exports.chromeStackLineParser,
    exports.geckoStackLineParser,
];
exports.defaultStackParser = (0, utils_1.createStackParser)(...exports.defaultStackLineParsers);
const extractSafariExtensionDetails = (func, filename) => {
    const isSafariExtension = func.indexOf('safari-extension') !== -1;
    const isSafariWebExtension = func.indexOf('safari-web-extension') !== -1;
    return isSafariExtension || isSafariWebExtension
        ? [
            func.indexOf('@') !== -1 ? func.split('@')[0] : utils_1.UNKNOWN_FUNCTION,
            isSafariExtension
                ? `safari-extension:${filename}`
                : `safari-web-extension:${filename}`,
        ]
        : [func, filename];
};
