// @hyperdx/instrumentation-exception/build/src/browser/integrations/browserapierrors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.browserApiErrorsIntegration = void 0;
const core_1 = require("@sentry/core");
const utils_1 = require("@sentry/utils");
const helpers_1 = require("../helpers");
const DEFAULT_EVENT_TARGET = [
    'EventTarget',
    'Window',
    'Node',
    'ApplicationCache',
    'AudioTrackList',
    'BroadcastChannel',
    'ChannelMergerNode',
    'CryptoOperation',
    'EventSource',
    'FileReader',
    'HTMLUnknownElement',
    'IDBDatabase',
    'IDBRequest',
    'IDBTransaction',
    'KeyOperation',
    'MediaController',
    'MessagePort',
    'ModalWindow',
    'Notification',
    'SVGElementInstance',
    'Screen',
    'SharedWorker',
    'TextTrack',
    'TextTrackCue',
    'TextTrackList',
    'WebSocket',
    'WebSocketWorker',
    'Worker',
    'XMLHttpRequest',
    'XMLHttpRequestEventTarget',
    'XMLHttpRequestUpload',
];
const INTEGRATION_NAME = 'BrowserApiErrors';
const _browserApiErrorsIntegration = ((options = {}) => {
    const _options = {
        XMLHttpRequest: true,
        eventTarget: true,
        requestAnimationFrame: true,
        setInterval: true,
        setTimeout: true,
        ...options,
    };
    return {
        name: INTEGRATION_NAME,
        setupOnce() {
            if (_options.setTimeout) {
                (0, utils_1.fill)(helpers_1.WINDOW, 'setTimeout', _wrapTimeFunction);
            }
            if (_options.setInterval) {
                (0, utils_1.fill)(helpers_1.WINDOW, 'setInterval', _wrapTimeFunction);
            }
            if (_options.requestAnimationFrame) {
                (0, utils_1.fill)(helpers_1.WINDOW, 'requestAnimationFrame', _wrapRAF);
            }
            if (_options.XMLHttpRequest && 'XMLHttpRequest' in helpers_1.WINDOW) {
                (0, utils_1.fill)(XMLHttpRequest.prototype, 'send', _wrapXHR);
            }
            const eventTargetOption = _options.eventTarget;
            if (eventTargetOption) {
                const eventTarget = Array.isArray(eventTargetOption)
                    ? eventTargetOption
                    : DEFAULT_EVENT_TARGET;
                eventTarget.forEach(_wrapEventTarget);
            }
        },
    };
});
exports.browserApiErrorsIntegration = (0, core_1.defineIntegration)(_browserApiErrorsIntegration);
function _wrapTimeFunction(original) {
    return function (...args) {
        const originalCallback = args[0];
        args[0] = (0, helpers_1.wrap)(originalCallback, {
            mechanism: {
                data: { function: (0, utils_1.getFunctionName)(original) },
                handled: false,
                type: 'instrument',
            },
        });
        return original.apply(this, args);
    };
}
function _wrapRAF(original) {
    return function (callback) {
        return original.apply(this, [
            (0, helpers_1.wrap)(callback, {
                mechanism: {
                    data: {
                        function: 'requestAnimationFrame',
                        handler: (0, utils_1.getFunctionName)(original),
                    },
                    handled: false,
                    type: 'instrument',
                },
            }),
        ]);
    };
}
function _wrapXHR(originalSend) {
    return function (...args) {
        const xhr = this;
        const xmlHttpRequestProps = [
            'onload',
            'onerror',
            'onprogress',
            'onreadystatechange',
        ];
        xmlHttpRequestProps.forEach((prop) => {
            if (prop in xhr && typeof xhr[prop] === 'function') {
                (0, utils_1.fill)(xhr, prop, function (original) {
                    const wrapOptions = {
                        mechanism: {
                            data: {
                                function: prop,
                                handler: (0, utils_1.getFunctionName)(original),
                            },
                            handled: false,
                            type: 'instrument',
                        },
                    };
                    const originalFunction = (0, utils_1.getOriginalFunction)(original);
                    if (originalFunction) {
                        wrapOptions.mechanism.data.handler =
                            (0, utils_1.getFunctionName)(originalFunction);
                    }
                    return (0, helpers_1.wrap)(original, wrapOptions);
                });
            }
        });
        return originalSend.apply(this, args);
    };
}
function _wrapEventTarget(target) {
    const globalObject = helpers_1.WINDOW;
    const proto = globalObject[target] && globalObject[target].prototype;
    if (!proto ||
        !proto.hasOwnProperty ||
        !proto.hasOwnProperty('addEventListener')) {
        return;
    }
    (0, utils_1.fill)(proto, 'addEventListener', function (original) {
        return function (eventName, fn, options) {
            try {
                if (typeof fn.handleEvent === 'function') {
                    fn.handleEvent = (0, helpers_1.wrap)(fn.handleEvent, {
                        mechanism: {
                            data: {
                                function: 'handleEvent',
                                handler: (0, utils_1.getFunctionName)(fn),
                                target,
                            },
                            handled: false,
                            type: 'instrument',
                        },
                    });
                }
            }
            catch (err) {
            }
            return original.apply(this, [
                eventName,
                (0, helpers_1.wrap)(fn, {
                    mechanism: {
                        data: {
                            function: 'addEventListener',
                            handler: (0, utils_1.getFunctionName)(fn),
                            target,
                        },
                        handled: false,
                        type: 'instrument',
                    },
                }),
                options,
            ]);
        };
    });
    (0, utils_1.fill)(proto, 'removeEventListener', function (originalRemoveEventListener) {
        return function (eventName, fn, options) {
            const wrappedEventHandler = fn;
            try {
                const originalEventHandler = wrappedEventHandler && wrappedEventHandler.__sentry_wrapped__;
                if (originalEventHandler) {
                    originalRemoveEventListener.call(this, eventName, originalEventHandler, options);
                }
            }
            catch (e) {
            }
            return originalRemoveEventListener.call(this, eventName, wrappedEventHandler, options);
        };
    });
}
