// @hyperdx/instrumentation-exception/build/src/browser/integrations/globalhandlers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.globalHandlersIntegration = void 0;
const core_1 = require("@sentry/core");
const api_1 = require("@opentelemetry/api");
const utils_1 = require("@sentry/utils");
const eventbuilder_1 = require("../eventbuilder");
const helpers_1 = require("../helpers");
const INTEGRATION_NAME = 'GlobalHandlers';
const _globalHandlersIntegration = ((options = {}) => {
    const _options = {
        onerror: true,
        onunhandledrejection: true,
        ...options,
    };
    return {
        name: INTEGRATION_NAME,
        setupOnce() {
            Error.stackTraceLimit = 50;
        },
        setup(client) {
            if (_options.onerror) {
                _installGlobalOnErrorHandler(client);
                globalHandlerLog('onerror');
            }
            if (_options.onunhandledrejection) {
                _installGlobalOnUnhandledRejectionHandler(client);
                globalHandlerLog('onunhandledrejection');
            }
        },
    };
});
exports.globalHandlersIntegration = (0, core_1.defineIntegration)(_globalHandlersIntegration);
function _installGlobalOnErrorHandler(client) {
    (0, utils_1.addGlobalErrorInstrumentationHandler)((data) => {
        const { stackParser, attachStacktrace } = getOptions();
        if ((0, core_1.getClient)() !== client || (0, helpers_1.shouldIgnoreOnError)()) {
            return;
        }
        const { msg, url, line, column, error } = data;
        const event = _enhanceEventWithInitialFrame((0, eventbuilder_1.eventFromUnknownInput)(stackParser, error || msg, undefined, attachStacktrace, false), url, line, column);
        event.level = 'error';
        (0, core_1.captureEvent)(event, {
            originalException: error,
            mechanism: {
                handled: false,
                type: 'onerror',
            },
        });
    });
}
function _installGlobalOnUnhandledRejectionHandler(client) {
    (0, utils_1.addGlobalUnhandledRejectionInstrumentationHandler)((e) => {
        const { stackParser, attachStacktrace } = getOptions();
        if ((0, core_1.getClient)() !== client || (0, helpers_1.shouldIgnoreOnError)()) {
            return;
        }
        const error = _getUnhandledRejectionError(e);
        const event = (0, utils_1.isPrimitive)(error)
            ? _eventFromRejectionWithPrimitive(error)
            : (0, eventbuilder_1.eventFromUnknownInput)(stackParser, error, undefined, attachStacktrace, true);
        event.level = 'error';
        (0, core_1.captureEvent)(event, {
            originalException: error,
            mechanism: {
                handled: false,
                type: 'onunhandledrejection',
            },
        });
    });
}
function _getUnhandledRejectionError(error) {
    if ((0, utils_1.isPrimitive)(error)) {
        return error;
    }
    try {
        if ('reason' in error) {
            return error.reason;
        }
        if ('detail' in error &&
            'reason' in error.detail) {
            return error.detail.reason;
        }
    }
    catch { }
    return error;
}
function _eventFromRejectionWithPrimitive(reason) {
    return {
        exception: {
            values: [
                {
                    type: 'UnhandledRejection',
                    value: `Non-Error promise rejection captured with value: ${String(reason)}`,
                },
            ],
        },
    };
}
function _enhanceEventWithInitialFrame(event, url, line, column) {
    const e = (event.exception = event.exception || {});
    const ev = (e.values = e.values || []);
    const ev0 = (ev[0] = ev[0] || {});
    const ev0s = (ev0.stacktrace = ev0.stacktrace || {});
    const ev0sf = (ev0s.frames = ev0s.frames || []);
    const colno = isNaN(parseInt(column, 10)) ? undefined : column;
    const lineno = isNaN(parseInt(line, 10)) ? undefined : line;
    const filename = (0, utils_1.isString)(url) && url.length > 0 ? url : (0, utils_1.getLocationHref)();
    if (ev0sf.length === 0) {
        ev0sf.push({
            colno,
            filename,
            function: utils_1.UNKNOWN_FUNCTION,
            in_app: true,
            lineno,
        });
    }
    return event;
}
function globalHandlerLog(type) {
    api_1.diag.debug(`Global Handler attached: ${type}`);
}
function getOptions() {
    const client = (0, core_1.getClient)();
    const options = (client && client.getOptions()) || {
        stackParser: () => [],
        attachStacktrace: false,
    };
    return options;
}
