// @hyperdx/instrumentation-exception/build/src/browser/integrations/reportingobserver.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.reportingObserverIntegration = void 0;
const core_1 = require("@sentry/core");
const utils_1 = require("@sentry/utils");
const WINDOW = utils_1.GLOBAL_OBJ;
const INTEGRATION_NAME = 'ReportingObserver';
const SETUP_CLIENTS = new WeakMap();
const _reportingObserverIntegration = ((options = {}) => {
    const types = options.types || ['crash', 'deprecation', 'intervention'];
    function handler(reports) {
        if (!SETUP_CLIENTS.has((0, core_1.getClient)())) {
            return;
        }
        for (const report of reports) {
            (0, core_1.withScope)((scope) => {
                scope.setExtra('url', report.url);
                const label = `ReportingObserver [${report.type}]`;
                let details = 'No details available';
                if (report.body) {
                    const plainBody = {};
                    for (const prop in report.body) {
                        plainBody[prop] = report.body[prop];
                    }
                    scope.setExtra('body', plainBody);
                    if (report.type === 'crash') {
                        const body = report.body;
                        details =
                            [body.crashId || '', body.reason || ''].join(' ').trim() ||
                                details;
                    }
                    else {
                        const body = report.body;
                        details = body.message || details;
                    }
                }
                (0, core_1.captureMessage)(`${label}: ${details}`);
            });
        }
    }
    return {
        name: INTEGRATION_NAME,
        setupOnce() {
            if (!(0, utils_1.supportsReportingObserver)()) {
                return;
            }
            const observer = new WINDOW.ReportingObserver(handler, {
                buffered: true,
                types,
            });
            observer.observe();
        },
        setup(client) {
            SETUP_CLIENTS.set(client, true);
        },
    };
});
exports.reportingObserverIntegration = (0, core_1.defineIntegration)(_reportingObserverIntegration);
